{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generation of non-sense and frame-shift mutated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bgvep.readers import Tabix\n",
    "import re\n",
    "from tqdm.notebook import trange, tqdm\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare CPTAC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load CPTAC dataset\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a3302539b4f48389ffec6089019c20d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=7.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Eliminate duplicated rna measures\n"
     ]
    }
   ],
   "source": [
    "####################\n",
    "#Load CPTAC dataset\n",
    "####################\n",
    "\n",
    "print('Load CPTAC dataset')\n",
    "\n",
    "CPTAC = ['BRCA', 'CCRCC', 'COAD', 'GBM', 'LUAD', 'OV', 'UCEC']\n",
    "cptac_df = pd.DataFrame(columns=['gene', 'cpct_aliquot', 'protein_expression', 'aliquot', 'sample',\n",
    "       'Type', 'median', 'stdev', 'norm_protein_expression', 'log2fpkm',\n",
    "       'log10fpkm', '#Uploaded_variation', 'Location', 'Allele', 'Feature',\n",
    "       'protein_mutation', 'Protein_position', 'Amino_acids', 'Consequence',\n",
    "       'Phenotype', 'Ubiquitinases_Mutated', 'Altered_E3_Ligases',\n",
    "       'Raw_Residual', 'Stability_Change', 'ABS_Stability_Change'])\n",
    "for cancer in tqdm(CPTAC, total=len(CPTAC)):\n",
    "    df = pd.read_csv(\"/workspace/projects/cptac_analysis/data/\"+cancer+\"/dataset_irls.gz\", sep='\\t')\n",
    "    Dataset = [cancer] * len(df)\n",
    "    df['Dataset'] = Dataset\n",
    "    frames = [cptac_df, df]\n",
    "    cptac_df = pd.concat(frames)\n",
    "cptac_df.drop_duplicates(keep='first',inplace=True) \n",
    "\n",
    "######################################################################################################\n",
    "#Calculate the mean of rna values for the repeated mutations (log2fpkm, log10fpkm, Stability_Change)\n",
    "######################################################################################################\n",
    "\n",
    "print('Eliminate duplicated rna measures')\n",
    "\n",
    "dupl_samples = ['C3L-00908', 'C3N-00545', 'C3N-01825']\n",
    "\n",
    "cptac_dupl_df = cptac_df[cptac_df['sample'].isin(dupl_samples)]\n",
    "df1 = cptac_dupl_df[cptac_dupl_df['Feature'].isnull()].groupby(['gene','sample'],as_index=False).mean()\n",
    "df2 = cptac_dupl_df[~cptac_dupl_df['Feature'].isnull()].groupby(['gene','Feature','#Uploaded_variation','Location'],as_index=False).mean()\n",
    "cptac_dupl2_df = pd.concat([df1,df2])\n",
    "cptac_dupl3_df = cptac_dupl_df.drop(['log2fpkm','log10fpkm','Raw_Residual','Stability_Change','ABS_Stability_Change'],axis=1)\n",
    "cptac_dupl4_df = pd.merge(cptac_dupl3_df,cptac_dupl2_df,how='left')\n",
    "cptac_dupl4_df.drop_duplicates(keep='first',inplace=True)\n",
    "cptac2_df = cptac_df[~(cptac_df['sample'].isin(dupl_samples))]\n",
    "cptac_df = pd.concat([cptac2_df,cptac_dupl4_df],ignore_index=True)\n",
    "\n",
    "#####################################################################################\n",
    "#Prepare cptac_df table for the merge (select stop_gained and frameshift_variant)\n",
    "#####################################################################################\n",
    "\n",
    "cptac_mut_df = cptac_df[['gene','Feature','#Uploaded_variation','Location','protein_mutation','Protein_position','Amino_acids','sample','Dataset']][(cptac_df['Phenotype']=='stop_gained')|(cptac_df['Phenotype']=='frameshift_variant')]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Prepare CCLE dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load CCLE dataset\n",
      "Eliminate duplicated protein measures\n"
     ]
    }
   ],
   "source": [
    "####################\n",
    "#Load CCLE dataset\n",
    "####################\n",
    "\n",
    "print('Load CCLE dataset')\n",
    "\n",
    "ccle_df = pd.read_csv(\"/workspace/projects/cptac_analysis/data/CCLE/dataset_irls.gz\", sep='\\t')\n",
    "\n",
    "#############################################################################################################################\n",
    "#Calculate the mean of protein values for the repeated mutations (protein_expression, norm_protein_expression, Raw_Residual, Stability_Change, ABS_Stability_Change)\n",
    "#############################################################################################################################\n",
    "\n",
    "print('Eliminate duplicated protein measures')\n",
    "\n",
    "df1 = ccle_df[ccle_df['Phenotype']=='WT'].groupby(['gene','sample','Phenotype'],as_index=False).mean()\n",
    "df2 = ccle_df[ccle_df['Phenotype']!='WT'].groupby(['gene','Feature','sample','#Uploaded_variation','Location','Phenotype'],as_index=False).mean()\n",
    "ccle2_df = pd.concat([df1,df2])\n",
    "ccle3_df = ccle_df.drop(['protein_expression', 'norm_protein_expression', 'Raw_Residual', 'Stability_Change', 'ABS_Stability_Change'],axis=1)\n",
    "ccle4_df = pd.merge(ccle3_df,ccle2_df,how='left')\n",
    "ccle_df = ccle4_df.drop_duplicates(keep='first')\n",
    "\n",
    "##################################################################################\n",
    "#Prepare cptac_df table for the merge(select stop_gained and frameshift_variant)\n",
    "##################################################################################\n",
    "\n",
    "ccle_mut_df = ccle_df[['gene','Feature','sample','#Uploaded_variation','Location','Phenotype']][(ccle_df['Phenotype']=='stop_gained')|(ccle_df['Phenotype']=='frameshift_variant')]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Import and prepare necessary tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Import tables with wt seq (BioMart) from canonical transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import wt sequences from canonical transcripts\n",
    "seq_df = pd.read_csv(\"/workspace/users/msanchezg/notebooks/ens_canonical_sequences.tsv\", sep=\"\\t\")\n",
    "\n",
    "#Merge mutations_df with seq_df\n",
    "seq_merge_df = seq_df[['ensembl_transcript_id','coding']]\n",
    "seq_merge_df = seq_merge_df.rename(columns={'ensembl_transcript_id':'Feature','coding':'seq_wt'})\n",
    "\n",
    "cptac_mut_df = pd.merge(cptac_mut_df, seq_merge_df, how='left')\n",
    "ccle_mut_df = pd.merge(ccle_mut_df, seq_merge_df, how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Tables with genes with ns and fs mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get genes containing ns or fs in cptac or ccle datasets\n",
    "genes_cptac_df = cptac_mut_df[['Feature']].drop_duplicates(subset='Feature',keep='first',ignore_index=True)\n",
    "genes_ccle_df = ccle_mut_df[['Feature']].drop_duplicates(subset='Feature',keep='first',ignore_index=True)\n",
    "genes_ns_fs_df = pd.concat([genes_cptac_df,genes_ccle_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Import CDS positions from BG-VEP "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cds_pos_bgvep (mut_df, positions_df):\n",
    "    '''Function to get cds positions for each mutation using bgvep annotation'''\n",
    "\n",
    "    positions_list = positions_df.values.tolist()\n",
    "\n",
    "    col_names = ['Feature','Chromosome','Position','Reference','Alternate','cDNA_position','CDS_position','Strand']\n",
    "\n",
    "    positions_list100 = positions_list[0:100]\n",
    "    cds_pos_df = pd.DataFrame(columns=col_names)\n",
    "    for position in tqdm(positions_list):\n",
    "        chrom = position[0]\n",
    "        pos = int(position[1])\n",
    "        with Tabix('hg38', '92') as reader:\n",
    "            for data in reader.get(chrom, pos, pos):\n",
    "                canonical = re.findall('YES', data[21])\n",
    "                if canonical !=[]:\n",
    "                    data2 =[]\n",
    "                    data2.append(data[5])\n",
    "                    data2.append(data[0])\n",
    "                    data2.append(data[1])\n",
    "                    data2.append(data[2])\n",
    "                    data2.append(data[3])\n",
    "                    data2.append(data[8])\n",
    "                    data2.append(data[9])\n",
    "                    data2.append(data[16])\n",
    "                    newRow = pd.DataFrame([data2], columns=col_names)\n",
    "                    cds_pos_df = cds_pos_df.append(newRow, ignore_index=True) \n",
    "    \n",
    "    return (cds_pos_df)         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cds_pos_table (mut_df):\n",
    "    '''Function to create a table with cds positions of the mutations'''\n",
    "    mut_df['chrom'] = mut_df['Location'].str.split(':').str[0]\n",
    "    mut_df['location2'] = mut_df['Location'].str.split(':').str[1]\n",
    "    mut_df['location2'] = mut_df['location2'].str.split('-').str[0]\n",
    "    positions_df = mut_df[['chrom','location2']]\n",
    "    cds_pos_df = cds_pos_bgvep (mut_df, positions_df)\n",
    "    cds_pos_df = cds_pos_df.rename(columns={'Position':'location2'})\n",
    "    cds_pos_df['location2'] = cds_pos_df['location2'].astype(str)\n",
    "    cds_pos_df['chrom'] = 'chr' + cds_pos_df['Chromosome'].astype(str)\n",
    "    cds_pos_df = cds_pos_df.drop(columns=['Chromosome','Alternate'])\n",
    "    cds_pos_df.drop_duplicates(keep='first',inplace=True)\n",
    "    return (cds_pos_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "de2fb17b808c4406837db3e73b4d9962",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=16421.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Add cds positions to datasets (takes a lot of time!)\n",
    "cptac_cds_pos_df = cds_pos_table (cptac_mut_df)\n",
    "ccle_cds_pos_df = cds_pos_table (cptac_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step is to save the tables XXXX_cds_pos_df, because the function **cds_pos_bgvep** (inside the function **cds_pos_table**) takes a lot of time, so we can run it only once and then load the saved tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save tables(to not compute it again, takes a lot of time)\n",
    "cptac_cds_pos_df.to_csv(r'cds_pos_nsfs_cptac.tsv', header = True, index = None, sep = '\\t')\n",
    "ccle_cds_pos_df.to_csv(r'cds_pos_nsfs_ccle.tsv', header = True, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load tables\n",
    "cptac_cds_pos_df = pd.read_csv('cds_pos_nsfs_ccle_bgveps.tsv', sep='\\t')\n",
    "ccle_cds_pos_df = pd.read_csv('cds_pos_nsfs_ccle_bgveps.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From here we can continue after uploading the saved XXXX_cds_pos_df tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Merge mutations_df with cds_pos_df\n",
    "cptac_mut_df = pd.merge(cptac_mut_df,cptac_cds_pos_df, how='left')\n",
    "ccle_mut_df = pd.merge(ccle_mut_df,ccle_cds_pos_df, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select mutations with CDS positions for in silico mutational process\n",
    "cptac_mut_df = cptac_mut_df[cptac_mut_df['CDS_position']!='-']\n",
    "cptac_mut_df = cptac_mut_df[~cptac_mut_df['CDS_position'].isnull()]\n",
    "ccle_mut_df = ccle_mut_df[ccle_mut_df['CDS_position']!='-']\n",
    "ccle_mut_df = ccle_mut_df[~ccle_mut_df['CDS_position'].isnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Create mutated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_seq (df):\n",
    "    '''Function to generate the mutated sequences with nonsense and frameshift alterations'''\n",
    "    \n",
    "    mutations_list = df[['seq_wt','CDS_position','REF','ALT','Reference','Strand','Feature','sample','Location','#Uploaded_variation','Phenotype']].values.tolist()\n",
    "\n",
    "    def fix_nt (nt, strand):\n",
    "        compl_dict = {'A':'T','T':'A','C':'G','G':'C'}\n",
    "        if strand == -1: \n",
    "            if len(nt)>1:\n",
    "                nt_list = list(nt)\n",
    "                nt2_list = []\n",
    "                for base in nt_list:\n",
    "                    nt2 = compl_dict[base]\n",
    "                    nt2_list.append(nt2)\n",
    "                nt2 = ''.join(nt2_list)\n",
    "            else:\n",
    "                nt2 = compl_dict[nt]\n",
    "        else:\n",
    "            nt2 = nt\n",
    "        return(nt2)\n",
    "\n",
    "    mut_check_list =[]\n",
    "    seq_mut_list =[]\n",
    "    for seq in mutations_list:\n",
    "        ref = seq[2]\n",
    "        alt = seq[3]\n",
    "        ref_b = seq[4]\n",
    "        strand = seq[5]\n",
    "        mutation = seq[10]\n",
    "        if seq[0] is np.nan:\n",
    "            mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan','nan','nan','nan','nan','nan','nan','nan','nan','nan','nan','nan']\n",
    "            seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq[0]]\n",
    "        else:\n",
    "            if (ref!='-')&(alt!='-'):\n",
    "                ref2 = fix_nt(ref, strand)\n",
    "                alt2 = fix_nt(alt, strand)\n",
    "                ref2_list = list(ref2)\n",
    "                cds_pos = int(seq[1])\n",
    "                seq_wt = seq[0]\n",
    "                if len(seq_wt)<cds_pos:\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,'nan',alt,alt2, len(seq_wt),'nan',len(ref2),len(alt2),'nan']\n",
    "                    seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'Short_seq_wt']\n",
    "                else:\n",
    "                    ref3 = seq_wt[cds_pos-1]\n",
    "                    if ref2_list[0] != ref3:\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt),'nan',len(ref2), len(alt2),'nan']\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'No_ref_match']\n",
    "                    else:\n",
    "                        seq_mut = seq_wt[0:(cds_pos-1)]+alt2+seq_wt[cds_pos:len(seq_wt)]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq_mut]\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],seq_wt[(cds_pos-2):(cds_pos+3)],seq_mut[(cds_pos-2):(cds_pos+3)],ref,ref2,ref3,alt,alt2,len(seq_wt), len(seq_mut), len(ref2), len(alt2), (len(seq_mut)-len(seq_wt))]\n",
    "            elif alt!='-':\n",
    "                ref2 = fix_nt(ref_b, strand)\n",
    "                alt2 = fix_nt(alt, strand)\n",
    "                cds_pos = int(seq[1])\n",
    "                seq_wt = seq[0]\n",
    "                if len(seq_wt)<cds_pos:\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,'nan',alt,alt2, len(seq_wt),'nan',len(ref2),len(alt2),'nan']\n",
    "                    seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'Short_seq_wt']\n",
    "                else:\n",
    "                    ref3 = seq_wt[cds_pos-1]\n",
    "                    if ref2 != ref3:\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt),'nan',len(ref2), len(alt2),'nan']\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'No_ref_match']\n",
    "                    else:\n",
    "                        seq_mut = seq_wt[0:(cds_pos-1)]+alt2+seq_wt[(cds_pos-1):len(seq_wt)]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq_mut]\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],seq_wt[(cds_pos-2):(cds_pos+3)],seq_mut[(cds_pos-2):(cds_pos+len(alt2)+3)],ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt), len(seq_mut), len(ref2), len(alt2), (len(seq_mut)-len(seq_wt))]\n",
    "            else:\n",
    "                ref2 = fix_nt(ref, strand)\n",
    "                alt2 = '-'\n",
    "                ref2_list = list(ref2)\n",
    "                cds_pos = int(seq[1])\n",
    "                seq_wt = seq[0]\n",
    "                if len(seq_wt)<cds_pos:\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref2,'nan',alt,alt2, len(seq_wt),'nan',len(ref2),len(alt2),'nan']\n",
    "                    seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'Short_seq_wt']\n",
    "                else:\n",
    "                    ref3 = seq_wt[cds_pos-1]\n",
    "                    if ref2_list[0] != ref3:\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt),'nan',len(ref2), len(alt2),'nan']\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'No_ref_match']\n",
    "                    else:\n",
    "                        seq_mut = seq_wt[0:(cds_pos-1)]+seq_wt[(cds_pos+len(ref2)-1):len(seq_wt)]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq_mut]\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],seq_wt[(cds_pos-2):(cds_pos+3)],seq_mut[(cds_pos-2):(cds_pos+3)],ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt), len(seq_mut), len(ref2), len(alt2), (len(seq_mut)-len(seq_wt))]\n",
    "\n",
    "        mut_check_list.append(mut_check)\n",
    "        seq_mut_list.append(seq_mut2)\n",
    "\n",
    "    mut_check_df = pd.DataFrame(mut_check_list, columns=['Feature','sample','Location','#Uploaded_cariation','seq_wt','seq_mut','ref','ref_bgveps','ref_at_strand','ref_at_seq','alt','alt_at_strand','length_seq_wt','length_seq_mut','length_ref','length_alt','length_seq_mut_minus_length_seq_wt'])\n",
    "    seq_mut_df = pd.DataFrame(seq_mut_list, columns=['Feature','sample','Location','#Uploaded_variation','Phenotype','REF','ALT','seq_mut'])\n",
    "    return (mut_check_df, seq_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate mutated sequences\n",
    "cptac_mutcheck_df, cptac_seq_mut_df = mutate_seq(cptac_mut_df)\n",
    "ccle_mutcheck_df, ccle_seq_mut_df = mutate_seq(ccle_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Table to check if in silico mutations have been done correctly\n",
    "cptac_mut_check_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Table to check if in silico mutations have been done correctly\n",
    "ccle_mut_check_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>sample</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Location1</th>\n",
       "      <th>...</th>\n",
       "      <th>Ins</th>\n",
       "      <th>REF2</th>\n",
       "      <th>ALT2</th>\n",
       "      <th>POS2</th>\n",
       "      <th>seq_wt</th>\n",
       "      <th>Reference</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>seq_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB7</td>\n",
       "      <td>ENST00000645829</td>\n",
       "      <td>11BR016__G__C</td>\n",
       "      <td>chrX:75060263</td>\n",
       "      <td>S669*</td>\n",
       "      <td>669</td>\n",
       "      <td>S/*</td>\n",
       "      <td>11BR016</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>chrX:75060263</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>75060263</td>\n",
       "      <td>ATGGCGCTGCTCGCGATGCATTCTTGGCGCTGGGCGGCCGCGGCGG...</td>\n",
       "      <td>G</td>\n",
       "      <td>2074</td>\n",
       "      <td>2006</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGCGCTGCTCGCGATGCATTCTTGGCGCTGGGCGGCCGCGGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCC4</td>\n",
       "      <td>ENST00000645237</td>\n",
       "      <td>20BR006__A__T</td>\n",
       "      <td>chr13:95166186</td>\n",
       "      <td>L669*</td>\n",
       "      <td>669</td>\n",
       "      <td>L/*</td>\n",
       "      <td>20BR006</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>chr13:95166186</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>95166186</td>\n",
       "      <td>ATGCTGCCCGTGTACCAGGAGGTGAAGCCCAACCCGCTGCAGGACG...</td>\n",
       "      <td>A</td>\n",
       "      <td>2138</td>\n",
       "      <td>2006</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGCTGCCCGTGTACCAGGAGGTGAAGCCCAACCCGCTGCAGGACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABI1</td>\n",
       "      <td>ENST00000376142</td>\n",
       "      <td>18BR003__G__A</td>\n",
       "      <td>chr10:26823179</td>\n",
       "      <td>Q82*</td>\n",
       "      <td>82</td>\n",
       "      <td>Q/*</td>\n",
       "      <td>18BR003</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>chr10:26823179</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>26823179</td>\n",
       "      <td>ATGGCAGAGCTGCAGATGTTACTAGAGGAGGAGATCCCGTCTGGCA...</td>\n",
       "      <td>G</td>\n",
       "      <td>316</td>\n",
       "      <td>244</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGCAGAGCTGCAGATGTTACTAGAGGAGGAGATCCCGTCTGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACACA</td>\n",
       "      <td>ENST00000616317</td>\n",
       "      <td>14BR008__ATACGTTTTGAAA__-</td>\n",
       "      <td>chr17:37274257-37274269</td>\n",
       "      <td>FSKRI311-315X</td>\n",
       "      <td>311-315</td>\n",
       "      <td>FSKRI/X</td>\n",
       "      <td>14BR008</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>chr17:37274257</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>ATACGTTTTGAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>37274257</td>\n",
       "      <td>ATGTGGTGGTCTACTCTGATGTCAATCTTGAGGGCTAGGTCTTTTT...</td>\n",
       "      <td>G</td>\n",
       "      <td>1426</td>\n",
       "      <td>945</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGTGGTGGTCTACTCTGATGTCAATCTTGAGGGCTAGGTCTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACBD6</td>\n",
       "      <td>ENST00000642319</td>\n",
       "      <td>11BR028__C__A</td>\n",
       "      <td>chr1:180397563</td>\n",
       "      <td>E206*</td>\n",
       "      <td>206</td>\n",
       "      <td>E/*</td>\n",
       "      <td>11BR028</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>chr1:180397563</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>180397563</td>\n",
       "      <td>ATGGCTTCATCATTCCTGCCCGCGGGGGCCATCACCGGCGACAGCG...</td>\n",
       "      <td>C</td>\n",
       "      <td>1304</td>\n",
       "      <td>616</td>\n",
       "      <td>-1</td>\n",
       "      <td>ATGGCTTCATCATTCCTGCCCGCGGGGGCCATCACCGGCGACAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16423</th>\n",
       "      <td>PTPRC</td>\n",
       "      <td>ENST00000442510</td>\n",
       "      <td>C3N-01825__-__AACTCCT</td>\n",
       "      <td>chr1:198748182-198748183</td>\n",
       "      <td>N974KTPX</td>\n",
       "      <td>974</td>\n",
       "      <td>N/KTPX</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>chr1:198748182</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>AACTCCT</td>\n",
       "      <td>198748182</td>\n",
       "      <td>ATGACCATGTATTTGTGGCTTAAACTCTTGGCATTTGGCTTTGCCT...</td>\n",
       "      <td>A</td>\n",
       "      <td>3095</td>\n",
       "      <td>2921</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGACCATGTATTTGTGGCTTAAACTCTTGGCATTTGGCTTTGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16424</th>\n",
       "      <td>RAB22A</td>\n",
       "      <td>ENST00000244040</td>\n",
       "      <td>C3N-01825__-__AAAAAAA</td>\n",
       "      <td>chr20:58354224-58354225</td>\n",
       "      <td>A149AKKX</td>\n",
       "      <td>149</td>\n",
       "      <td>A/AKKX</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>chr20:58354224</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>58354224</td>\n",
       "      <td>ATGGCGCTGAGGGAGCTCAAAGTGTGTCTGCTCGGGGATACAGGTG...</td>\n",
       "      <td>C</td>\n",
       "      <td>727</td>\n",
       "      <td>446</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGGCGCTGAGGGAGCTCAAAGTGTGTCTGCTCGGGGATACAGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16425</th>\n",
       "      <td>SLC7A2</td>\n",
       "      <td>ENST00000004531</td>\n",
       "      <td>C3N-01825__-__ATTCA</td>\n",
       "      <td>chr8:17548817-17548818</td>\n",
       "      <td>-264-265IX</td>\n",
       "      <td>264-265</td>\n",
       "      <td>-/IX</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>chr8:17548817</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>ATTCA</td>\n",
       "      <td>17548817</td>\n",
       "      <td>ATGAAGATAGAAACAAGTGGTTATAACTCAGACAAACTAATTTGTC...</td>\n",
       "      <td>T</td>\n",
       "      <td>840</td>\n",
       "      <td>792</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGAAGATAGAAACAAGTGGTTATAACTCAGACAAACTAATTTGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16426</th>\n",
       "      <td>TAOK1</td>\n",
       "      <td>ENST00000261716</td>\n",
       "      <td>C3N-01825__-__TTCATTG</td>\n",
       "      <td>chr17:29510944-29510945</td>\n",
       "      <td>-552-553FIX</td>\n",
       "      <td>552-553</td>\n",
       "      <td>-/FIX</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>chr17:29510944</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>TTCATTG</td>\n",
       "      <td>29510944</td>\n",
       "      <td>ATGCCATCAACTAACAGAGCAGGCAGCCTGAAGGACCCTGAAATTG...</td>\n",
       "      <td>C</td>\n",
       "      <td>2175</td>\n",
       "      <td>1656</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGCCATCAACTAACAGAGCAGGCAGCCTGAAGGACCCTGAAATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16427</th>\n",
       "      <td>UBN2</td>\n",
       "      <td>ENST00000473989</td>\n",
       "      <td>C3N-01825__-__TGGGTCT</td>\n",
       "      <td>chr7:139237094-139237095</td>\n",
       "      <td>-186-187WVX</td>\n",
       "      <td>186-187</td>\n",
       "      <td>-/WVX</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>chr7:139237094</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>TGGGTCT</td>\n",
       "      <td>139237094</td>\n",
       "      <td>ATGGCGGAGCCGCGCAGAGTAGCGTTCATTAGCTTGTCACCGGTGC...</td>\n",
       "      <td>A</td>\n",
       "      <td>558</td>\n",
       "      <td>558</td>\n",
       "      <td>1</td>\n",
       "      <td>ATGGCGGAGCCGCGCAGAGTAGCGTTCATTAGCTTGTCACCGGTGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16428 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene          Feature        #Uploaded_variation  \\\n",
       "0       ABCB7  ENST00000645829              11BR016__G__C   \n",
       "1       ABCC4  ENST00000645237              20BR006__A__T   \n",
       "2        ABI1  ENST00000376142              18BR003__G__A   \n",
       "3       ACACA  ENST00000616317  14BR008__ATACGTTTTGAAA__-   \n",
       "4       ACBD6  ENST00000642319              11BR028__C__A   \n",
       "...       ...              ...                        ...   \n",
       "16423   PTPRC  ENST00000442510      C3N-01825__-__AACTCCT   \n",
       "16424  RAB22A  ENST00000244040      C3N-01825__-__AAAAAAA   \n",
       "16425  SLC7A2  ENST00000004531        C3N-01825__-__ATTCA   \n",
       "16426   TAOK1  ENST00000261716      C3N-01825__-__TTCATTG   \n",
       "16427    UBN2  ENST00000473989      C3N-01825__-__TGGGTCT   \n",
       "\n",
       "                       Location protein_mutation Protein_position Amino_acids  \\\n",
       "0                 chrX:75060263            S669*              669         S/*   \n",
       "1                chr13:95166186            L669*              669         L/*   \n",
       "2                chr10:26823179             Q82*               82         Q/*   \n",
       "3       chr17:37274257-37274269    FSKRI311-315X          311-315     FSKRI/X   \n",
       "4                chr1:180397563            E206*              206         E/*   \n",
       "...                         ...              ...              ...         ...   \n",
       "16423  chr1:198748182-198748183         N974KTPX              974      N/KTPX   \n",
       "16424   chr20:58354224-58354225         A149AKKX              149      A/AKKX   \n",
       "16425    chr8:17548817-17548818       -264-265IX          264-265        -/IX   \n",
       "16426   chr17:29510944-29510945      -552-553FIX          552-553       -/FIX   \n",
       "16427  chr7:139237094-139237095      -186-187WVX          186-187       -/WVX   \n",
       "\n",
       "          sample Dataset       Location1  ... Ins           REF2     ALT2  \\\n",
       "0        11BR016    BRCA   chrX:75060263  ...   0              G        C   \n",
       "1        20BR006    BRCA  chr13:95166186  ...   0              A        T   \n",
       "2        18BR003    BRCA  chr10:26823179  ...   0              G        A   \n",
       "3        14BR008    BRCA  chr17:37274257  ...   0  ATACGTTTTGAAA        -   \n",
       "4        11BR028    BRCA  chr1:180397563  ...   0              C        A   \n",
       "...          ...     ...             ...  ...  ..            ...      ...   \n",
       "16423  C3N-01825    UCEC  chr1:198748182  ...  -1              -  AACTCCT   \n",
       "16424  C3N-01825    UCEC  chr20:58354224  ...  -1              -  AAAAAAA   \n",
       "16425  C3N-01825    UCEC   chr8:17548817  ...  -1              -    ATTCA   \n",
       "16426  C3N-01825    UCEC  chr17:29510944  ...  -1              -  TTCATTG   \n",
       "16427  C3N-01825    UCEC  chr7:139237094  ...  -1              -  TGGGTCT   \n",
       "\n",
       "            POS2                                             seq_wt  \\\n",
       "0       75060263  ATGGCGCTGCTCGCGATGCATTCTTGGCGCTGGGCGGCCGCGGCGG...   \n",
       "1       95166186  ATGCTGCCCGTGTACCAGGAGGTGAAGCCCAACCCGCTGCAGGACG...   \n",
       "2       26823179  ATGGCAGAGCTGCAGATGTTACTAGAGGAGGAGATCCCGTCTGGCA...   \n",
       "3       37274257  ATGTGGTGGTCTACTCTGATGTCAATCTTGAGGGCTAGGTCTTTTT...   \n",
       "4      180397563  ATGGCTTCATCATTCCTGCCCGCGGGGGCCATCACCGGCGACAGCG...   \n",
       "...          ...                                                ...   \n",
       "16423  198748182  ATGACCATGTATTTGTGGCTTAAACTCTTGGCATTTGGCTTTGCCT...   \n",
       "16424   58354224  ATGGCGCTGAGGGAGCTCAAAGTGTGTCTGCTCGGGGATACAGGTG...   \n",
       "16425   17548817  ATGAAGATAGAAACAAGTGGTTATAACTCAGACAAACTAATTTGTC...   \n",
       "16426   29510944  ATGCCATCAACTAACAGAGCAGGCAGCCTGAAGGACCCTGAAATTG...   \n",
       "16427  139237094  ATGGCGGAGCCGCGCAGAGTAGCGTTCATTAGCTTGTCACCGGTGC...   \n",
       "\n",
       "       Reference  cDNA_position CDS_position Strand  \\\n",
       "0              G           2074         2006     -1   \n",
       "1              A           2138         2006     -1   \n",
       "2              G            316          244     -1   \n",
       "3              G           1426          945     -1   \n",
       "4              C           1304          616     -1   \n",
       "...          ...            ...          ...    ...   \n",
       "16423          A           3095         2921      1   \n",
       "16424          C            727          446      1   \n",
       "16425          T            840          792      1   \n",
       "16426          C           2175         1656      1   \n",
       "16427          A            558          558      1   \n",
       "\n",
       "                                                 seq_mut  \n",
       "0      ATGGCGCTGCTCGCGATGCATTCTTGGCGCTGGGCGGCCGCGGCGG...  \n",
       "1      ATGCTGCCCGTGTACCAGGAGGTGAAGCCCAACCCGCTGCAGGACG...  \n",
       "2      ATGGCAGAGCTGCAGATGTTACTAGAGGAGGAGATCCCGTCTGGCA...  \n",
       "3      ATGTGGTGGTCTACTCTGATGTCAATCTTGAGGGCTAGGTCTTTTT...  \n",
       "4      ATGGCTTCATCATTCCTGCCCGCGGGGGCCATCACCGGCGACAGCG...  \n",
       "...                                                  ...  \n",
       "16423  ATGACCATGTATTTGTGGCTTAAACTCTTGGCATTTGGCTTTGCCT...  \n",
       "16424  ATGGCGCTGAGGGAGCTCAAAGTGTGTCTGCTCGGGGATACAGGTG...  \n",
       "16425  ATGAAGATAGAAACAAGTGGTTATAACTCAGACAAACTAATTTGTC...  \n",
       "16426  ATGCCATCAACTAACAGAGCAGGCAGCCTGAAGGACCCTGAAATTG...  \n",
       "16427  ATGGCGGAGCCGCGCAGAGTAGCGTTCATTAGCTTGTCACCGGTGC...  \n",
       "\n",
       "[16428 rows x 26 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Add mutated sequence to cptac_mut_df\n",
    "cptac_mut_df = pd.merge(cptac_mut_df,cptac_seq_mut_df,how='left')\n",
    "ccle_mut_df = pd.merge(ccle_mut_df,ccle_seq_mut_df,how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_code = {'GAA': 'E', 'CGA': 'R', 'GTG': 'V', 'TAA': '*', 'CGT': 'R', 'ATA': 'I', 'GAC': 'D', 'TCG': 'S', \n",
    "                  'GAT': 'D', 'ATG': 'M', 'CTG': 'L', 'CTA': 'L', 'TAC': 'Y', 'GGA': 'G', 'CGG': 'R', 'AGC': 'S', \n",
    "                  'TCT': 'S', 'TGA': '*', 'AAA': 'K', 'ACC': 'T', 'ACA': 'T', 'TGC': 'C', 'AAG': 'K', 'GTC': 'V', \n",
    "                  'TCC': 'S', 'ACT': 'T', 'AGA': 'R', 'CTT': 'L', 'GCC': 'A', 'GTA': 'V', 'TAG': '*', 'CAA': 'Q', \n",
    "                  'CAC': 'H', 'GCT': 'A', 'TTA': 'L', 'CAT': 'H', 'CGC': 'R', 'TTC': 'F', 'ATT': 'I', 'GGC': 'G', \n",
    "                  'CAG': 'Q', 'AAC': 'N', 'CCC': 'P', 'GTT': 'V', 'AGG': 'R', 'TGT': 'C', 'CCG': 'P', 'GGG': 'G', \n",
    "                  'ATC': 'I', 'TTT': 'F', 'AAT': 'N', 'TCA': 'S', 'GAG': 'E', 'CCA': 'P', 'GCA': 'A', 'TAT': 'Y', \n",
    "                  'GGT': 'G', 'TGG': 'W', 'GCG': 'A', 'CTC': 'L', 'TTG': 'L', 'CCT': 'P', 'ACG': 'T', 'AGT': 'S'}\n",
    "\n",
    "def translation(nt_seq_list):\n",
    "    '''Function to translate in silico a list of cdna sequences'''\n",
    "    prot_seq_list = []\n",
    "    for nt_seq in nt_seq_list:\n",
    "        if (nt_seq[0] is np.nan)|(nt_seq[0] == 'Sequence unavailable')|(nt_seq[0] == 'No_ref_match')|(nt_seq[0] == 'Short_seq_wt'):\n",
    "            prot_seq_ids = [nt_seq[1],nt_seq[2],nt_seq[3],nt_seq[0]]\n",
    "            prot_seq_list.append(prot_seq_ids)\n",
    "        else:\n",
    "            nt_seq_triplets = [nt_seq[0][i:i+3] for i in range(0, len(nt_seq[0]), 3)]\n",
    "            prot_seq = ''\n",
    "\n",
    "            for triplet in nt_seq_triplets:\n",
    "                if len(triplet) == 3:\n",
    "                    if 'N' in triplet:\n",
    "                        prot_seq += 'X' \n",
    "                    else:\n",
    "                        prot_seq += genetic_code[triplet]\n",
    "                        if '*' in prot_seq:\n",
    "                            break\n",
    "                else:\n",
    "                    break\n",
    "            prot_seq_ids = [nt_seq[1],nt_seq[2],nt_seq[3],prot_seq]\n",
    "            prot_seq_list.append(prot_seq_ids)\n",
    "    return(prot_seq_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translation_mut_table (df):\n",
    "    '''Function to prepare a table with translated mutated sequences'''\n",
    "    seq_mut_list = df[['seq_mut','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    prot_seq_mut_list = translation(seq_mut_list)\n",
    "    prot_seq_mut_df = pd.DataFrame(prot_seq_mut_list, columns=['Feature','#Uploaded_variation','Location','prot_seq'])\n",
    "    prot_mut_df = pd.merge(df,prot_seq_mut_df,how='left')\n",
    "    return (prot_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Translate sequences CCLE\n",
    "cptac_prot_mut_df = translation_mut_table(cptac_mut_df)\n",
    "ccle_prot_mut_df = translation_mut_table(ccle_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translation_wt_table (df,seq_df):\n",
    "    '''Function to prepare a table with translated wt sequences'''\n",
    "    #Prepare cptac_df table for the merge\n",
    "    wt_df = df[['gene','Feature','#Uploaded_variation','Location','protein_mutation','Protein_position','Amino_acids','sample','Dataset']][df['Phenotype']!='WT']\n",
    "    wt_df.drop_duplicates(keep='first',inplace=True)\n",
    "\n",
    "    #Merge seq with cptac table for all proteins\n",
    "    seq_wt_df = seq_df[['ensembl_transcript_id','coding']]\n",
    "    seq_wt_df.drop_duplicates(keep='first',inplace=True)\n",
    "    seq_wt_df = seq_wt_df.rename(columns={'ensembl_transcript_id':'Feature','coding':'seq_wt'})\n",
    "\n",
    "    wt_df = pd.merge(wt_df, seq_wt_df, how='left')\n",
    "\n",
    "    #Translate wt sequences\n",
    "    seq_wt_list = wt_df[['seq_wt','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    prot_seq_wt_list = translation(seq_wt_list)\n",
    "    prot_wt_df = pd.DataFrame(prot_seq_wt_list, columns=['Feature','#Uploaded_variation','Location','prot_seq'])\n",
    "    return (prot_wt_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a table qith translated sequences\n",
    "cptac_prot_wt_df = translation_wt_table (cptac_df,seq_df)\n",
    "ccle_prot_wt_df = translation_wt_table (ccle_df,seq_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Search for c-term degrons in the generated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_cdegron(prot_list, cdegron_motif):\n",
    "    \"\"\"This function finds all proteins containing c-terminal degrons (cdegrons)\"\"\"\n",
    "    cdegron_seq_list = [] \n",
    "    for mutation in prot_list:\n",
    "        seq = mutation[0]\n",
    "        splitted = re.split('\\*', str(seq))\n",
    "        seq2 = splitted[0]\n",
    "        motif_re = cdegron_motif +'$'\n",
    "        motif_re = motif_re.replace('X', '.')\n",
    "        find_cdegron_motif = re.findall(motif_re, seq2)\n",
    "        if find_cdegron_motif != []:\n",
    "            mutation_list = []\n",
    "            mutation_list.append(mutation[1])\n",
    "            mutation_list.append(mutation[2])\n",
    "            mutation_list.append(mutation[3])\n",
    "            mutation_list.append(seq)\n",
    "            mutation_list.append(True)\n",
    "            cdegron_seq_list.append(mutation_list)\n",
    "    cdegron_df = pd.DataFrame(cdegron_seq_list,columns=['Feature','#Uploaded_variation','Location','prot_seq',cdegron_motif])\n",
    "    return cdegron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "cdegron_motifs = ['GG', 'RG', 'PG', 'XR', 'RXXG', 'EE', 'RXX', 'VX', 'AX', 'A']\n",
    "\n",
    "def find_cdegron_list(prot_list, cdegron_motifs):\n",
    "    \"\"\"This function finds all cdegron motifs provided in a list\"\"\"\n",
    "    col_names=['Feature','#Uploaded_variation','Location','prot_seq']\n",
    "    all_cdegron_df = pd.DataFrame(columns=col_names)\n",
    "    for motif in cdegron_motifs:\n",
    "        cdegron_df = find_cdegron(prot_list, motif)\n",
    "        all_cdegron_df = pd.concat([all_cdegron_df,cdegron_df])\n",
    "    all_cdegron_df = all_cdegron_df.fillna(False)\n",
    "    dupl = all_cdegron_df[all_cdegron_df[['Feature','#Uploaded_variation','Location']].duplicated()].values.tolist()\n",
    "    df2 = pd.DataFrame()\n",
    "    for seq in dupl:\n",
    "        df1 = all_cdegron_df[(all_cdegron_df['Feature']==seq[0])&(all_cdegron_df['#Uploaded_variation']==seq[1])&(all_cdegron_df['Location']==seq[2])].groupby(['Feature','#Uploaded_variation','Location','prot_seq'], as_index=False).sum()\n",
    "        df2 = pd.concat([df2,df1])\n",
    "    all_cdegron_df.drop_duplicates(subset=['Feature','#Uploaded_variation','Location','prot_seq'], keep=False, inplace=True)\n",
    "    df2.drop_duplicates(keep='first', inplace=True)\n",
    "    all_cdegron_df = pd.concat([all_cdegron_df,df2],ignore_index=True)\n",
    "    return all_cdegron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cdegron_table (prot_mut_df, prot_wt_df, cdegron_motifs,df,seq_df)\n",
    "    '''This function creates a table with c-degrons annotated per sequence'''\n",
    "    #Search cdegron motifs in cptac_mut_df\n",
    "    prot_seq_mut_list = prot_mut_df[['prot_seq_mut','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    cdegron_mut_df = find_cdegron_list(prot_seq_mut_list,cdegron_motifs)\n",
    "\n",
    "    #Add proteins without cterm degrons to the table\n",
    "    prot_seq_mut_df = cptac_prot_mut_df[['gene','Feature','#Uploaded_variation','Location','prot_seq']]\n",
    "    prot_seq_mut_df = prot_seq_mut_df.rename(columns={'prot_seq_mut':'prot_seq'})\n",
    "    cdegron_mut_df = pd.merge(prot_seq_mut_df,cdegron_mut_df,how='left')\n",
    "    cdegron_mut_df = cdegron_mut_df.fillna(False)\n",
    "    \n",
    "    #Find wt proteins with cterm degrons\n",
    "    prot_wt_list = prot_wt_df[['prot_seq','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    cdegron_wt_df = find_cdegron_list(prot_wt_list,cdegron_motifs)\n",
    "\n",
    "    #Add proteins without cterm degrons to the table\n",
    "    cdegron_wt_df = pd.merge(prot_wt_df,cdegron_wt_df,how='left')\n",
    "    cdegron_wt_df = cdegron_wt_df.fillna(False)\n",
    "\n",
    "    #Add 'gene' and 'Phenotype' to cdegron_wt_df\n",
    "    id_df = cptac_df[['gene','Feature']][cptac_df['Phenotype']!='WT']\n",
    "    id_df.drop_duplicates(keep='first',inplace=True)\n",
    "    cdegron_wt_df = pd.merge(cdegron_wt_df,id_df,how='left')\n",
    "\n",
    "    cdegron_wt2_df = cdegron_wt_df.drop(columns=['Feature','#Uploaded_variation','Location'])\n",
    "    phenotype_wt = 'WT'\n",
    "    cdegron_wt2_df['Phenotype'] = phenotype_wt\n",
    "\n",
    "    #Add phenotype to cdegron_mut_df\n",
    "    phenotype_mut = df[['Feature','#Uploaded_variation','Location','Phenotype']][(df['Phenotype']=='stop_gained')|(df['Phenotype']=='frameshift_variant')|(df['Phenotype']=='stop_lost')]\n",
    "    cdegron_mut_df = pd.merge(cdegron_mut_df,phenotype_mut,how='left')\n",
    "    cdegron_mut_df = pd.merge(cdegron_mut_df,id_df,how='left')\n",
    "\n",
    "    #concat wt and mut\n",
    "    cdegron_df = pd.concat([cdegron_wt2_df,cdegron_mut_df])\n",
    "    cdegron_df.drop_duplicates(keep='first',inplace=True) \n",
    "    return (cdegron_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create tables with c-degron annotations\n",
    "cptac_cdegron_df = cdegron_table (cptac_prot_mut_df, cptac_prot_wt_df, cdegron_motifs, cptac_df, seq_df)\n",
    "ccle_cdegron_df = cdegron_table (ccle_prot_mut_df, ccle_prot_wt_df, cdegron_motifs, ccle_df, seq_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save table with cdegron annotation\n",
    "cptac_cdegron_df.to_csv(r'cdegron_wtnsfs_cptac.tsv', header = True, index = None, sep = '\\t')\n",
    "ccle_cdegron_df.to_csv(r'cdegron_wtnsfs_ccle.tsv', header = True, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Get c-terminal amino acid (last aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_last_aa(prot_list, aa):\n",
    "    \"\"\"This function finds all proteins containing c-terminal degrons (cdegrons)\"\"\"\n",
    "    aa_seq_list = [] \n",
    "    for mutation in prot_list:\n",
    "        seq = mutation[0]\n",
    "        splitted = re.split('\\*', str(seq))\n",
    "        seq2 = splitted[0]\n",
    "        aa_re = aa +'$'\n",
    "        find_aa = re.findall(aa_re, seq2)\n",
    "        if find_aa != []:\n",
    "            mutation_list = []\n",
    "            mutation_list.append(mutation[1])\n",
    "            mutation_list.append(mutation[2])\n",
    "            mutation_list.append(mutation[3])\n",
    "            mutation_list.append(seq)\n",
    "            mutation_list.append(aa)\n",
    "            aa_seq_list.append(mutation_list)\n",
    "    aa_df = pd.DataFrame(aa_seq_list,columns=['Feature','#Uploaded_variation','Location','prot_seq','last_aa'])\n",
    "    return aa_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_list = ['A','C','D','E','F','G','H','I','K','Q','L','M','N','P','R','S','T','V','W','Y']\n",
    "\n",
    "def get_last_aa_list(prot_list, aa_list):\n",
    "    \"\"\"This function finds all cdegron motifs provided in a list\"\"\"\n",
    "    col_names=['Feature','#Uploaded_variation','Location','prot_seq']\n",
    "    all_aa_df = pd.DataFrame()\n",
    "    for aa in aa_list:\n",
    "        aa_df = get_last_aa(prot_list, aa)\n",
    "        all_aa_df = pd.concat([all_aa_df,aa_df],ignore_index=True)\n",
    "    return all_aa_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def last_aa_table (prot_mut_df,aa_list, prot_wt_df, df)\n",
    "    '''This function annotates all c-terminal amino acids (last aa) from mutated and wt sequences'''\n",
    "    \n",
    "    prot_seq_mut_list = prot_mut_df[['prot_seq_mut','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    aa_mut_df = get_last_aa_list(prot_seq_mut_list,aa_list)\n",
    "\n",
    "    #Get last aa from wt sequences\n",
    "    prot_wt_list = prot_wt_df[['prot_seq','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    aa_wt_df = get_last_aa_list(prot_wt_list,aa_list)\n",
    "\n",
    "    #Add 'gene' and 'Phenotype'\n",
    "    id_df = cptac_df[['gene','Feature']][cptac_df['Phenotype']!='WT']\n",
    "    id_df.drop_duplicates(keep='first',inplace=True)\n",
    "    aa_wt_df = pd.merge(aa_wt_df,id_df,how='left')\n",
    "\n",
    "    aa_wt2_df = aa_wt_df.drop(columns=['Feature','#Uploaded_variation','Location'])\n",
    "    phenotype_wt = 'WT'\n",
    "    aa_wt2_df['Phenotype'] = phenotype_wt\n",
    "\n",
    "    phenotype_mut = df[['Feature','#Uploaded_variation','Location','Phenotype']][(df['Phenotype']=='stop_gained')|(df['Phenotype']=='frameshift_variant')]\n",
    "    aa_mut_df = pd.merge(aa_mut_df,phenotype_mut,how='left')\n",
    "    aa_mut_df = pd.merge(aa_mut_df,id_df,how='left')\n",
    "\n",
    "    #concat wt and mut\n",
    "    aa_df = pd.concat([aa_wt2_df,aa_mut_df])\n",
    "    aa_df.drop_duplicates(keep='first',inplace=True) \n",
    "    return (aa_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create tables with last aa annotation\n",
    "cptac_aa_df = last_aa_table (cptac_prot_mut_df, aa_list, cptac_prot_wt_df, cptac_df)\n",
    "ccle_aa_df = last_aa_table (ccle_prot_mut_df, aa_list, ccle_prot_wt_df, ccle_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save table with last aa annotation\n",
    "cptac_aa_df.to_csv(r'last_aa_cptac.tsv', header = True, index = None, sep = '\\t')\n",
    "ccle_aa_df.to_csv(r'last_aa_ccle.tsv', header = True, index = None, sep = '\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:cptac]",
   "language": "python",
   "name": "conda-env-cptac-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
