{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generation of nonsense and frameshift mutated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from bgvep.readers import Tabix\n",
    "import re\n",
    "from tqdm.notebook import trange, tqdm\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare CPTAC dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load CPTAC dataset\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9c82e4aa86534758b69d7f3853f3a201",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=7.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Eliminate duplicated rna measures\n",
      "Select mutated sequences\n"
     ]
    }
   ],
   "source": [
    "####################\n",
    "#Load CPTAC dataset\n",
    "####################\n",
    "\n",
    "print('Load CPTAC dataset')\n",
    "\n",
    "CPTAC = ['BRCA', 'CCRCC', 'COAD', 'GBM', 'LUAD', 'OV', 'UCEC']\n",
    "cptac_df = pd.DataFrame(columns=['gene', 'cpct_aliquot', 'protein_expression', 'aliquot', 'sample',\n",
    "       'Type', 'median', 'stdev', 'norm_protein_expression', 'log2fpkm',\n",
    "       'log10fpkm', '#Uploaded_variation', 'Location', 'Allele', 'Feature',\n",
    "       'protein_mutation', 'Protein_position', 'Amino_acids', 'Consequence',\n",
    "       'Phenotype', 'Ubiquitinases_Mutated', 'Altered_E3_Ligases',\n",
    "       'Raw_Residual', 'Stability_Change', 'ABS_Stability_Change'])\n",
    "for cancer in tqdm(CPTAC, total=len(CPTAC)):\n",
    "    df = pd.read_csv(\"/workspace/projects/cptac_analysis/data/\"+cancer+\"/dataset_irls.gz\", sep='\\t')\n",
    "    Dataset = [cancer] * len(df)\n",
    "    df['Dataset'] = Dataset\n",
    "    frames = [cptac_df, df]\n",
    "    cptac_df = pd.concat(frames)\n",
    "cptac_df.drop_duplicates(keep='first',inplace=True) \n",
    "\n",
    "cptac_df['ID'] = cptac_df['gene'].astype(str)+cptac_df['Feature'].astype(str)+cptac_df['sample'].astype(str)+cptac_df['#Uploaded_variation'].astype(str)+cptac_df['Location'].astype(str)\n",
    "\n",
    "######################################################################################################\n",
    "#Calculate the mean of rna values for the repeated mutations (log2fpkm, log10fpkm, Stability_Change)\n",
    "######################################################################################################\n",
    "\n",
    "print('Eliminate duplicated rna measures')\n",
    "\n",
    "dupl_samples = ['C3L-00908', 'C3N-00545', 'C3N-01825']\n",
    "\n",
    "cptac_dupl_df = cptac_df[cptac_df['sample'].isin(dupl_samples)]\n",
    "df1 = cptac_dupl_df[cptac_dupl_df['Feature'].isnull()].groupby(['gene','sample'],as_index=False).mean()\n",
    "df2 = cptac_dupl_df[~cptac_dupl_df['Feature'].isnull()].groupby(['gene','Feature','#Uploaded_variation','Location'],as_index=False).mean()\n",
    "cptac_dupl2_df = pd.concat([df1,df2])\n",
    "cptac_dupl3_df = cptac_dupl_df.drop(['log2fpkm','log10fpkm','Raw_Residual','Stability_Change','ABS_Stability_Change'],axis=1)\n",
    "cptac_dupl4_df = pd.merge(cptac_dupl3_df,cptac_dupl2_df,how='left')\n",
    "cptac_dupl4_df.drop_duplicates(keep='first',inplace=True)\n",
    "cptac2_df = cptac_df[~(cptac_df['sample'].isin(dupl_samples))]\n",
    "cptac_df = pd.concat([cptac2_df,cptac_dupl4_df],ignore_index=True)\n",
    "\n",
    "###########################################\n",
    "#Select stop_gained and frameshift_variant\n",
    "###########################################\n",
    "\n",
    "print('Select mutated sequences')\n",
    "\n",
    "cptac_mut_df = cptac_df[['gene','Feature','sample','#Uploaded_variation','Location','Phenotype','Dataset','ID']][(cptac_df['Phenotype']=='stop_gained')|(cptac_df['Phenotype']=='frameshift_variant')]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Prepare CCLE dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load CCLE dataset\n",
      "Eliminate duplicated protein measures\n",
      "Select mutated sequences\n"
     ]
    }
   ],
   "source": [
    "####################\n",
    "#Load CCLE dataset\n",
    "####################\n",
    "\n",
    "print('Load CCLE dataset')\n",
    "\n",
    "ccle_df = pd.read_csv(\"/workspace/projects/cptac_analysis/data/CCLE/dataset_irls.gz\", sep='\\t')\n",
    "\n",
    "ccle_df['ID'] = ccle_df['gene'].astype(str)+ccle_df['Feature'].astype(str)+ccle_df['sample'].astype(str)+ccle_df['#Uploaded_variation'].astype(str)+ccle_df['Location'].astype(str)\n",
    "\n",
    "#############################################################################################################################\n",
    "#Calculate the mean of protein values for the repeated mutations (protein_expression, norm_protein_expression, Raw_Residual, Stability_Change, ABS_Stability_Change)\n",
    "#############################################################################################################################\n",
    "\n",
    "print('Eliminate duplicated protein measures')\n",
    "\n",
    "df1 = ccle_df[ccle_df['Phenotype']=='WT'].groupby(['gene','sample','Phenotype'],as_index=False).mean()\n",
    "df2 = ccle_df[ccle_df['Phenotype']!='WT'].groupby(['gene','Feature','sample','#Uploaded_variation','Location','Phenotype'],as_index=False).mean()\n",
    "ccle2_df = pd.concat([df1,df2])\n",
    "ccle3_df = ccle_df.drop(['protein_expression', 'norm_protein_expression', 'Raw_Residual', 'Stability_Change', 'ABS_Stability_Change'],axis=1)\n",
    "ccle4_df = pd.merge(ccle3_df,ccle2_df,how='left')\n",
    "ccle_df = ccle4_df.drop_duplicates(keep='first')\n",
    "\n",
    "##################################################################################\n",
    "#Prepare cptac_df table for the merge(select stop_gained and frameshift_variant)\n",
    "##################################################################################\n",
    "\n",
    "print('Select mutated sequences')\n",
    "\n",
    "ccle_mut_df = ccle_df[['gene','Feature','sample','#Uploaded_variation','Location','Phenotype','ID']][(ccle_df['Phenotype']=='stop_gained')|(ccle_df['Phenotype']=='frameshift_variant')]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Import and prepare necessary tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Import tables with wt seq (BioMart) from canonical transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import wt sequences from canonical transcripts\n",
    "seq_df = pd.read_csv(\"/workspace/users/msanchezg/notebooks/ens_canonical_sequences.tsv\", sep=\"\\t\")\n",
    "\n",
    "#Merge mutations_df with seq_df\n",
    "seq_merge_df = seq_df[['ensembl_transcript_id','coding']]\n",
    "seq_merge_df = seq_merge_df.rename(columns={'ensembl_transcript_id':'Feature','coding':'seq_wt'})\n",
    "\n",
    "cptac_mut_df = pd.merge(cptac_mut_df, seq_merge_df, how='left')\n",
    "ccle_mut_df = pd.merge(ccle_mut_df, seq_merge_df, how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Import CDS positions from BG-VEP "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step is to save the tables XXXX_cds_pos_df, because the function **cds_pos_bgvep** (inside the function **cds_pos_table**) takes a lot of time, so we run it only once and then load the saved tables for posterior analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cds_pos_bgvep (mut_df, positions_df):\n",
    "    '''Function to get cds positions for each mutation using bgvep annotation'''\n",
    "\n",
    "    positions_list = positions_df.values.tolist()\n",
    "\n",
    "    col_names = ['Feature','Chromosome','Position','Reference','Alternate','cDNA_position','CDS_position','Strand']\n",
    "\n",
    "    positions_list100 = positions_list[0:100]\n",
    "    cds_pos_df = pd.DataFrame(columns=col_names)\n",
    "    for position in tqdm(positions_list):\n",
    "        chrom = position[0]\n",
    "        pos = int(position[1])\n",
    "        with Tabix('hg38', '92') as reader:\n",
    "            for data in reader.get(chrom, pos, pos):\n",
    "                canonical = re.findall('YES', data[21])\n",
    "                if canonical !=[]:\n",
    "                    data2 =[]\n",
    "                    data2.append(data[5])\n",
    "                    data2.append(data[0])\n",
    "                    data2.append(data[1])\n",
    "                    data2.append(data[2])\n",
    "                    data2.append(data[3])\n",
    "                    data2.append(data[8])\n",
    "                    data2.append(data[9])\n",
    "                    data2.append(data[16])\n",
    "                    newRow = pd.DataFrame([data2], columns=col_names)\n",
    "                    cds_pos_df = cds_pos_df.append(newRow, ignore_index=True) \n",
    "    \n",
    "    return (cds_pos_df)         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cds_pos_table (mut_df):\n",
    "    '''Function to create a table with cds positions of the mutations'''\n",
    "    mut_df['chrom'] = mut_df['Location'].str.split(':').str[0]\n",
    "    mut_df['location2'] = mut_df['Location'].str.split(':').str[1]\n",
    "    mut_df['location2'] = mut_df['location2'].str.split('-').str[0]\n",
    "    positions_df = mut_df[['chrom','location2']]\n",
    "    cds_pos_df = cds_pos_bgvep (mut_df, positions_df)\n",
    "    cds_pos_df = cds_pos_df.rename(columns={'Position':'location2'})\n",
    "    cds_pos_df['location2'] = cds_pos_df['location2'].astype(str)\n",
    "    cds_pos_df['chrom'] = 'chr' + cds_pos_df['Chromosome'].astype(str)\n",
    "    cds_pos_df = cds_pos_df.drop(columns=['Chromosome','Alternate'])\n",
    "    cds_pos_df.drop_duplicates(keep='first',inplace=True)\n",
    "    return (cds_pos_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1fcfba6d54b14c93b1882f296629354f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=14932.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#Add cds positions to datasets (takes a lot of time, 1-5h)\n",
    "cptac_cds_pos_df = cds_pos_table (cptac_mut_df)\n",
    "ccle_cds_pos_df = cds_pos_table (ccle_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save tables\n",
    "cptac_cds_pos_df.to_csv(r'cds_pos_nsfs_cptac.tsv', header = True, index = None, sep = '\\t')\n",
    "ccle_cds_pos_df.to_csv(r'cds_pos_nsfs_ccle.tsv', header = True, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From here we can continue after uploading the saved XXXX_cds_pos_df tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load tables\n",
    "cptac_cds_pos_df = pd.read_csv('cds_pos_nsfs_cptac.tsv', sep='\\t')\n",
    "ccle_cds_pos_df = pd.read_csv('cds_pos_nsfs_ccle.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Merge mutations_df with cds_pos_df\n",
    "\n",
    "cptac_cds_pos_df['location2'] = cptac_cds_pos_df['location2'].astype(str)\n",
    "cptac_mut_df['chrom'] = cptac_mut_df['Location'].str.split(':').str[0]\n",
    "cptac_mut_df['location2'] = cptac_mut_df['Location'].str.split(':').str[1]\n",
    "cptac_mut_df['location2'] = cptac_mut_df['location2'].str.split('-').str[0]\n",
    "cptac_mut_df = pd.merge(cptac_mut_df,cptac_cds_pos_df, how='left')\n",
    "\n",
    "ccle_cds_pos_df['location2'] = ccle_cds_pos_df['location2'].astype(str)\n",
    "ccle_mut_df['chrom'] = ccle_mut_df['Location'].str.split(':').str[0]\n",
    "ccle_mut_df['location2'] = ccle_mut_df['Location'].str.split(':').str[1]\n",
    "ccle_mut_df['location2'] = ccle_mut_df['location2'].str.split('-').str[0]\n",
    "ccle_mut_df = pd.merge(ccle_mut_df,ccle_cds_pos_df, how='left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select mutations with CDS positions for in silico mutational process\n",
    "cptac_mut_df = cptac_mut_df[cptac_mut_df['CDS_position']!='-']\n",
    "cptac_mut_df = cptac_mut_df[~cptac_mut_df['CDS_position'].isnull()]\n",
    "ccle_mut_df = ccle_mut_df[ccle_mut_df['CDS_position']!='-']\n",
    "ccle_mut_df = ccle_mut_df[~ccle_mut_df['CDS_position'].isnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Create mutated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_seq (df):\n",
    "    '''Function to generate the mutated sequences with nonsense and frameshift alterations'''\n",
    "    \n",
    "    df['REF'] = df['#Uploaded_variation'].str.split('__').str[1]\n",
    "    df['ALT'] = df['#Uploaded_variation'].str.split('__').str[2]\n",
    "    \n",
    "    mutations_list = df[['seq_wt','CDS_position','REF','ALT','Reference','Strand','Feature','sample','Location','#Uploaded_variation','Phenotype','ID']].values.tolist()\n",
    "\n",
    "    def fix_nt (nt, strand):\n",
    "        compl_dict = {'A':'T','T':'A','C':'G','G':'C'}\n",
    "        if strand == -1: \n",
    "            if len(nt)>1:\n",
    "                nt_list = list(nt)\n",
    "                nt2_list = []\n",
    "                for base in nt_list:\n",
    "                    nt2 = compl_dict[base]\n",
    "                    nt2_list.append(nt2)\n",
    "                nt2 = ''.join(nt2_list)\n",
    "            else:\n",
    "                nt2 = compl_dict[nt]\n",
    "        else:\n",
    "            nt2 = nt\n",
    "        return(nt2)\n",
    "\n",
    "    mut_check_list =[]\n",
    "    seq_mut_list =[]\n",
    "    for seq in mutations_list:\n",
    "        ref = seq[2]\n",
    "        alt = seq[3]\n",
    "        ref_b = seq[4]\n",
    "        strand = seq[5]\n",
    "        mutation = seq[10]\n",
    "        if seq[0] is np.nan:\n",
    "            mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan','nan','nan','nan','nan','nan','nan','nan','nan','nan','nan','nan',seq[11]]\n",
    "            seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq[0],seq[11]]\n",
    "        else:\n",
    "            if (ref!='-')&(alt!='-'):\n",
    "                ref2 = fix_nt(ref, strand)\n",
    "                alt2 = fix_nt(alt, strand)\n",
    "                ref2_list = list(ref2)\n",
    "                cds_pos = int(seq[1])\n",
    "                seq_wt = seq[0]\n",
    "                if len(seq_wt)<cds_pos:\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,'nan',alt,alt2, len(seq_wt),'nan',len(ref2),len(alt2),'nan',seq[11]]\n",
    "                    seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'Short_seq_wt',seq[11]]\n",
    "                else:\n",
    "                    ref3 = seq_wt[cds_pos-1]\n",
    "                    if ref2_list[0] != ref3:\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt),'nan',len(ref2), len(alt2),'nan',seq[11]]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'No_ref_match',seq[11]]\n",
    "                    else:\n",
    "                        seq_mut = seq_wt[0:(cds_pos-1)]+alt2+seq_wt[cds_pos:len(seq_wt)]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq_mut,seq[11]]\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],seq_wt[(cds_pos-2):(cds_pos+3)],seq_mut[(cds_pos-2):(cds_pos+3)],ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt), len(seq_mut), len(ref2), len(alt2), (len(seq_mut)-len(seq_wt)), seq[11]]\n",
    "            elif alt!='-':\n",
    "                ref2 = fix_nt(ref_b, strand)\n",
    "                alt2 = fix_nt(alt, strand)\n",
    "                cds_pos = int(seq[1])\n",
    "                seq_wt = seq[0]\n",
    "                if len(seq_wt)<cds_pos:\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,'nan',alt,alt2, len(seq_wt),'nan',len(ref2),len(alt2),'nan',seq[11]]\n",
    "                    seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'Short_seq_wt',seq[11]]\n",
    "                else:\n",
    "                    ref3 = seq_wt[cds_pos-1]\n",
    "                    if ref2 != ref3:\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt),'nan',len(ref2), len(alt2),'nan',seq[11]]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'No_ref_match',seq[11]]\n",
    "                    else:\n",
    "                        seq_mut = seq_wt[0:(cds_pos-1)]+alt2+seq_wt[(cds_pos-1):len(seq_wt)]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq_mut,seq[11]]\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],seq_wt[(cds_pos-2):(cds_pos+3)],seq_mut[(cds_pos-2):(cds_pos+len(alt2)+3)],ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt), len(seq_mut), len(ref2), len(alt2), (len(seq_mut)-len(seq_wt)),seq[11]]\n",
    "            else:\n",
    "                ref2 = fix_nt(ref, strand)\n",
    "                alt2 = '-'\n",
    "                ref2_list = list(ref2)\n",
    "                cds_pos = int(seq[1])\n",
    "                seq_wt = seq[0]\n",
    "                if len(seq_wt)<cds_pos:\n",
    "                    mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref2,'nan',alt,alt2, len(seq_wt),'nan',len(ref2),len(alt2),'nan',seq[11]]\n",
    "                    seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'Short_seq_wt',seq[11]]\n",
    "                else:\n",
    "                    ref3 = seq_wt[cds_pos-1]\n",
    "                    if ref2_list[0] != ref3:\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],'nan','nan',ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt),'nan',len(ref2), len(alt2),'nan',seq[11]]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,'No_ref_match',seq[11]]\n",
    "                    else:\n",
    "                        seq_mut = seq_wt[0:(cds_pos-1)]+seq_wt[(cds_pos+len(ref2)-1):len(seq_wt)]\n",
    "                        seq_mut2 = [seq[6],seq[7],seq[8],seq[9],seq[10],ref,alt,seq_mut,seq[11]]\n",
    "                        mut_check = [seq[6],seq[7],seq[8],seq[9],seq_wt[(cds_pos-2):(cds_pos+3)],seq_mut[(cds_pos-2):(cds_pos+3)],ref,ref_b,ref2,ref3,alt,alt2,len(seq_wt), len(seq_mut), len(ref2), len(alt2), (len(seq_mut)-len(seq_wt)),seq[11]]\n",
    "\n",
    "        mut_check_list.append(mut_check)\n",
    "        seq_mut_list.append(seq_mut2)\n",
    "\n",
    "    mut_check_df = pd.DataFrame(mut_check_list, columns=['Feature','sample','Location','#Uploaded_cariation','seq_wt','seq_mut','ref','ref_bgveps','ref_at_strand','ref_at_seq','alt','alt_at_strand','length_seq_wt','length_seq_mut','length_ref','length_alt','length_seq_mut_minus_length_seq_wt','ID'])\n",
    "    seq_mut_df = pd.DataFrame(seq_mut_list, columns=['Feature','sample','Location','#Uploaded_variation','Phenotype','REF','ALT','seq_mut','ID'])\n",
    "    return (mut_check_df, seq_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate mutated sequences\n",
    "cptac_mutcheck_df, cptac_seq_mut_df = mutate_seq(cptac_mut_df)\n",
    "ccle_mutcheck_df, ccle_seq_mut_df = mutate_seq(ccle_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>sample</th>\n",
       "      <th>Location</th>\n",
       "      <th>#Uploaded_cariation</th>\n",
       "      <th>seq_wt</th>\n",
       "      <th>seq_mut</th>\n",
       "      <th>ref</th>\n",
       "      <th>ref_bgveps</th>\n",
       "      <th>ref_at_strand</th>\n",
       "      <th>ref_at_seq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_at_strand</th>\n",
       "      <th>length_seq_wt</th>\n",
       "      <th>length_seq_mut</th>\n",
       "      <th>length_ref</th>\n",
       "      <th>length_alt</th>\n",
       "      <th>length_seq_mut_minus_length_seq_wt</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000645829</td>\n",
       "      <td>11BR016</td>\n",
       "      <td>chrX:75060263</td>\n",
       "      <td>11BR016__G__C</td>\n",
       "      <td>GCACA</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2274</td>\n",
       "      <td>2274</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ABCB7ENST0000064582911BR01611BR016__G__CchrX:7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000645237</td>\n",
       "      <td>20BR006</td>\n",
       "      <td>chr13:95166186</td>\n",
       "      <td>20BR006__A__T</td>\n",
       "      <td>TTGAA</td>\n",
       "      <td>TAGAA</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>3978</td>\n",
       "      <td>3978</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ABCC4ENST0000064523720BR00620BR006__A__Tchr13:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000376142</td>\n",
       "      <td>18BR003</td>\n",
       "      <td>chr10:26823179</td>\n",
       "      <td>18BR003__G__A</td>\n",
       "      <td>TCAGC</td>\n",
       "      <td>TTAGC</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1527</td>\n",
       "      <td>1527</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ABI1ENST0000037614218BR00318BR003__G__Achr10:2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000616317</td>\n",
       "      <td>14BR008</td>\n",
       "      <td>chr17:37274257-37274269</td>\n",
       "      <td>14BR008__ATACGTTTTGAAA__-</td>\n",
       "      <td>ATCTT</td>\n",
       "      <td>ACCAG</td>\n",
       "      <td>ATACGTTTTGAAA</td>\n",
       "      <td>A</td>\n",
       "      <td>TATGCAAAACTTT</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>7152</td>\n",
       "      <td>7139</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>-13</td>\n",
       "      <td>ACACAENST0000061631714BR00814BR008__ATACGTTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000642319</td>\n",
       "      <td>11BR028</td>\n",
       "      <td>chr1:180397563</td>\n",
       "      <td>11BR028__C__A</td>\n",
       "      <td>GGAAC</td>\n",
       "      <td>GTAAC</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>849</td>\n",
       "      <td>849</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ACBD6ENST0000064231911BR02811BR028__C__Achr1:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16396</th>\n",
       "      <td>ENST00000442510</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>chr1:198748182-198748183</td>\n",
       "      <td>C3N-01825__-__AACTCCT</td>\n",
       "      <td>AATTC</td>\n",
       "      <td>AAACTCCTATTC</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AACTCCT</td>\n",
       "      <td>AACTCCT</td>\n",
       "      <td>3921</td>\n",
       "      <td>3928</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>PTPRCENST00000442510C3N-01825C3N-01825__-__AAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16397</th>\n",
       "      <td>ENST00000244040</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>chr20:58354224-58354225</td>\n",
       "      <td>C3N-01825__-__AAAAAAA</td>\n",
       "      <td>GCAAA</td>\n",
       "      <td>GAAAAAAACAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>585</td>\n",
       "      <td>592</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>RAB22AENST00000244040C3N-01825C3N-01825__-__AA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16398</th>\n",
       "      <td>ENST00000004531</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>chr8:17548817-17548818</td>\n",
       "      <td>C3N-01825__-__ATTCA</td>\n",
       "      <td>TTCTC</td>\n",
       "      <td>TATTCATCTC</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>ATTCA</td>\n",
       "      <td>ATTCA</td>\n",
       "      <td>2097</td>\n",
       "      <td>2102</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>SLC7A2ENST00000004531C3N-01825C3N-01825__-__AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16399</th>\n",
       "      <td>ENST00000261716</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>chr17:29510944-29510945</td>\n",
       "      <td>C3N-01825__-__TTCATTG</td>\n",
       "      <td>TCGAG</td>\n",
       "      <td>TTTCATTGCGAG</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>TTCATTG</td>\n",
       "      <td>TTCATTG</td>\n",
       "      <td>3006</td>\n",
       "      <td>3013</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>TAOK1ENST00000261716C3N-01825C3N-01825__-__TTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16400</th>\n",
       "      <td>ENST00000473989</td>\n",
       "      <td>C3N-01825</td>\n",
       "      <td>chr7:139237094-139237095</td>\n",
       "      <td>C3N-01825__-__TGGGTCT</td>\n",
       "      <td>AATAT</td>\n",
       "      <td>ATGGGTCTATAT</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>TGGGTCT</td>\n",
       "      <td>TGGGTCT</td>\n",
       "      <td>4044</td>\n",
       "      <td>4051</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>UBN2ENST00000473989C3N-01825C3N-01825__-__TGGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16401 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Feature     sample                  Location  \\\n",
       "0      ENST00000645829    11BR016             chrX:75060263   \n",
       "1      ENST00000645237    20BR006            chr13:95166186   \n",
       "2      ENST00000376142    18BR003            chr10:26823179   \n",
       "3      ENST00000616317    14BR008   chr17:37274257-37274269   \n",
       "4      ENST00000642319    11BR028            chr1:180397563   \n",
       "...                ...        ...                       ...   \n",
       "16396  ENST00000442510  C3N-01825  chr1:198748182-198748183   \n",
       "16397  ENST00000244040  C3N-01825   chr20:58354224-58354225   \n",
       "16398  ENST00000004531  C3N-01825    chr8:17548817-17548818   \n",
       "16399  ENST00000261716  C3N-01825   chr17:29510944-29510945   \n",
       "16400  ENST00000473989  C3N-01825  chr7:139237094-139237095   \n",
       "\n",
       "             #Uploaded_cariation seq_wt       seq_mut            ref  \\\n",
       "0                  11BR016__G__C  GCACA         GGACA              G   \n",
       "1                  20BR006__A__T  TTGAA         TAGAA              A   \n",
       "2                  18BR003__G__A  TCAGC         TTAGC              G   \n",
       "3      14BR008__ATACGTTTTGAAA__-  ATCTT         ACCAG  ATACGTTTTGAAA   \n",
       "4                  11BR028__C__A  GGAAC         GTAAC              C   \n",
       "...                          ...    ...           ...            ...   \n",
       "16396      C3N-01825__-__AACTCCT  AATTC  AAACTCCTATTC              -   \n",
       "16397      C3N-01825__-__AAAAAAA  GCAAA  GAAAAAAACAAA              -   \n",
       "16398        C3N-01825__-__ATTCA  TTCTC    TATTCATCTC              -   \n",
       "16399      C3N-01825__-__TTCATTG  TCGAG  TTTCATTGCGAG              -   \n",
       "16400      C3N-01825__-__TGGGTCT  AATAT  ATGGGTCTATAT              -   \n",
       "\n",
       "      ref_bgveps  ref_at_strand ref_at_seq      alt alt_at_strand  \\\n",
       "0              G              C          C        C             G   \n",
       "1              A              T          T        T             A   \n",
       "2              G              C          C        A             T   \n",
       "3              A  TATGCAAAACTTT          T        -             -   \n",
       "4              C              G          G        A             T   \n",
       "...          ...            ...        ...      ...           ...   \n",
       "16396          A              A          A  AACTCCT       AACTCCT   \n",
       "16397          C              C          C  AAAAAAA       AAAAAAA   \n",
       "16398          T              T          T    ATTCA         ATTCA   \n",
       "16399          C              C          C  TTCATTG       TTCATTG   \n",
       "16400          A              A          A  TGGGTCT       TGGGTCT   \n",
       "\n",
       "      length_seq_wt length_seq_mut length_ref length_alt  \\\n",
       "0              2274           2274          1          1   \n",
       "1              3978           3978          1          1   \n",
       "2              1527           1527          1          1   \n",
       "3              7152           7139         13          1   \n",
       "4               849            849          1          1   \n",
       "...             ...            ...        ...        ...   \n",
       "16396          3921           3928          1          7   \n",
       "16397           585            592          1          7   \n",
       "16398          2097           2102          1          5   \n",
       "16399          3006           3013          1          7   \n",
       "16400          4044           4051          1          7   \n",
       "\n",
       "      length_seq_mut_minus_length_seq_wt  \\\n",
       "0                                      0   \n",
       "1                                      0   \n",
       "2                                      0   \n",
       "3                                    -13   \n",
       "4                                      0   \n",
       "...                                  ...   \n",
       "16396                                  7   \n",
       "16397                                  7   \n",
       "16398                                  5   \n",
       "16399                                  7   \n",
       "16400                                  7   \n",
       "\n",
       "                                                      ID  \n",
       "0      ABCB7ENST0000064582911BR01611BR016__G__CchrX:7...  \n",
       "1      ABCC4ENST0000064523720BR00620BR006__A__Tchr13:...  \n",
       "2      ABI1ENST0000037614218BR00318BR003__G__Achr10:2...  \n",
       "3      ACACAENST0000061631714BR00814BR008__ATACGTTTTG...  \n",
       "4      ACBD6ENST0000064231911BR02811BR028__C__Achr1:1...  \n",
       "...                                                  ...  \n",
       "16396  PTPRCENST00000442510C3N-01825C3N-01825__-__AAC...  \n",
       "16397  RAB22AENST00000244040C3N-01825C3N-01825__-__AA...  \n",
       "16398  SLC7A2ENST00000004531C3N-01825C3N-01825__-__AT...  \n",
       "16399  TAOK1ENST00000261716C3N-01825C3N-01825__-__TTC...  \n",
       "16400  UBN2ENST00000473989C3N-01825C3N-01825__-__TGGG...  \n",
       "\n",
       "[16401 rows x 18 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Table to check if in silico mutations have been done correctly\n",
    "cptac_mutcheck_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>sample</th>\n",
       "      <th>Location</th>\n",
       "      <th>#Uploaded_cariation</th>\n",
       "      <th>seq_wt</th>\n",
       "      <th>seq_mut</th>\n",
       "      <th>ref</th>\n",
       "      <th>ref_bgveps</th>\n",
       "      <th>ref_at_strand</th>\n",
       "      <th>ref_at_seq</th>\n",
       "      <th>alt</th>\n",
       "      <th>alt_at_strand</th>\n",
       "      <th>length_seq_wt</th>\n",
       "      <th>length_seq_mut</th>\n",
       "      <th>length_ref</th>\n",
       "      <th>length_alt</th>\n",
       "      <th>length_seq_mut_minus_length_seq_wt</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000262461</td>\n",
       "      <td>MCF7_BREAST</td>\n",
       "      <td>chr5:128180952-128180953</td>\n",
       "      <td>ACH-000019__-__A</td>\n",
       "      <td>TTCAT</td>\n",
       "      <td>TATCAT</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>3639</td>\n",
       "      <td>3640</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SLC12A2ENST00000262461MCF7_BREASTACH-000019__-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000262461</td>\n",
       "      <td>GB1_CENTRAL_NERVOUS_SYSTEM</td>\n",
       "      <td>chr5:128184383-128184384</td>\n",
       "      <td>ACH-000738__AA__-</td>\n",
       "      <td>GAAAT</td>\n",
       "      <td>GATCA</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>3639</td>\n",
       "      <td>3637</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-2</td>\n",
       "      <td>SLC12A2ENST00000262461GB1_CENTRAL_NERVOUS_SYST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000262461</td>\n",
       "      <td>SNGM_ENDOMETRIUM</td>\n",
       "      <td>chr5:128178638</td>\n",
       "      <td>ACH-000974__A__-</td>\n",
       "      <td>GAAAA</td>\n",
       "      <td>GAAAA</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>3639</td>\n",
       "      <td>3638</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>SLC12A2ENST00000262461SNGM_ENDOMETRIUMACH-0009...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000262461</td>\n",
       "      <td>SKES1_BONE</td>\n",
       "      <td>chr5:128131080</td>\n",
       "      <td>ACH-000087__T__-</td>\n",
       "      <td>ATTTA</td>\n",
       "      <td>ATTAA</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>3639</td>\n",
       "      <td>3638</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>SLC12A2ENST00000262461SKES1_BONEACH-000087__T_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000262461</td>\n",
       "      <td>MFE319_ENDOMETRIUM</td>\n",
       "      <td>chr5:128135792-128135793</td>\n",
       "      <td>ACH-000988__-__T</td>\n",
       "      <td>GGTTT</td>\n",
       "      <td>GTGTTT</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>3639</td>\n",
       "      <td>3640</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>SLC12A2ENST00000262461MFE319_ENDOMETRIUMACH-00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14911</th>\n",
       "      <td>ENST00000273371</td>\n",
       "      <td>NCIH446_LUNG</td>\n",
       "      <td>chr3:119608918</td>\n",
       "      <td>ACH-000800__G__T</td>\n",
       "      <td>CGAGA</td>\n",
       "      <td>CTAGA</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>1371</td>\n",
       "      <td>1371</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>PLA1AENST00000273371NCIH446_LUNGACH-000800__G_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14912</th>\n",
       "      <td>ENST00000378076</td>\n",
       "      <td>NCIH650_LUNG</td>\n",
       "      <td>chr10:15688006</td>\n",
       "      <td>ACH-000945__C__A</td>\n",
       "      <td>AGAAC</td>\n",
       "      <td>ATAAC</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>3192</td>\n",
       "      <td>3192</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ITGA8ENST00000378076NCIH650_LUNGACH-000945__C_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14913</th>\n",
       "      <td>ENST00000222598</td>\n",
       "      <td>HEC59_ENDOMETRIUM</td>\n",
       "      <td>chr7:97022295-97022296</td>\n",
       "      <td>ACH-000994__AA__-</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>DLX5ENST00000222598HEC59_ENDOMETRIUMACH-000994...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14914</th>\n",
       "      <td>ENST00000222598</td>\n",
       "      <td>HEC59_ENDOMETRIUM</td>\n",
       "      <td>chr7:97024466</td>\n",
       "      <td>ACH-000994__C__-</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>DLX5ENST00000222598HEC59_ENDOMETRIUMACH-000994...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14915</th>\n",
       "      <td>ENST00000331075</td>\n",
       "      <td>HEC59_ENDOMETRIUM</td>\n",
       "      <td>chr22:31125775</td>\n",
       "      <td>ACH-000994__C__T</td>\n",
       "      <td>CCGAT</td>\n",
       "      <td>CTGAT</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>3021</td>\n",
       "      <td>3021</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>INPP5JENST00000331075HEC59_ENDOMETRIUMACH-0009...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14916 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Feature                      sample                  Location  \\\n",
       "0      ENST00000262461                 MCF7_BREAST  chr5:128180952-128180953   \n",
       "1      ENST00000262461  GB1_CENTRAL_NERVOUS_SYSTEM  chr5:128184383-128184384   \n",
       "2      ENST00000262461            SNGM_ENDOMETRIUM            chr5:128178638   \n",
       "3      ENST00000262461                  SKES1_BONE            chr5:128131080   \n",
       "4      ENST00000262461          MFE319_ENDOMETRIUM  chr5:128135792-128135793   \n",
       "...                ...                         ...                       ...   \n",
       "14911  ENST00000273371                NCIH446_LUNG            chr3:119608918   \n",
       "14912  ENST00000378076                NCIH650_LUNG            chr10:15688006   \n",
       "14913  ENST00000222598           HEC59_ENDOMETRIUM    chr7:97022295-97022296   \n",
       "14914  ENST00000222598           HEC59_ENDOMETRIUM             chr7:97024466   \n",
       "14915  ENST00000331075           HEC59_ENDOMETRIUM            chr22:31125775   \n",
       "\n",
       "      #Uploaded_cariation seq_wt seq_mut  ref ref_bgveps ref_at_strand  \\\n",
       "0        ACH-000019__-__A  TTCAT  TATCAT    -          T             T   \n",
       "1       ACH-000738__AA__-  GAAAT   GATCA   AA          A            AA   \n",
       "2        ACH-000974__A__-  GAAAA   GAAAA    A          A             A   \n",
       "3        ACH-000087__T__-  ATTTA   ATTAA    T          T             T   \n",
       "4        ACH-000988__-__T  GGTTT  GTGTTT    -          G             G   \n",
       "...                   ...    ...     ...  ...        ...           ...   \n",
       "14911    ACH-000800__G__T  CGAGA   CTAGA    G          G             G   \n",
       "14912    ACH-000945__C__A  AGAAC   ATAAC    C          C             G   \n",
       "14913   ACH-000994__AA__-    nan     nan  nan        nan           nan   \n",
       "14914    ACH-000994__C__-    nan     nan  nan        nan           nan   \n",
       "14915    ACH-000994__C__T  CCGAT   CTGAT    C          C             C   \n",
       "\n",
       "      ref_at_seq  alt alt_at_strand length_seq_wt length_seq_mut length_ref  \\\n",
       "0              T    A             A          3639           3640          1   \n",
       "1              A    -             -          3639           3637          2   \n",
       "2              A    -             -          3639           3638          1   \n",
       "3              T    -             -          3639           3638          1   \n",
       "4              G    T             T          3639           3640          1   \n",
       "...          ...  ...           ...           ...            ...        ...   \n",
       "14911          G    T             T          1371           1371          1   \n",
       "14912          G    A             T          3192           3192          1   \n",
       "14913        nan  nan           nan           nan            nan        nan   \n",
       "14914        nan  nan           nan           nan            nan        nan   \n",
       "14915          C    T             T          3021           3021          1   \n",
       "\n",
       "      length_alt length_seq_mut_minus_length_seq_wt  \\\n",
       "0              1                                  1   \n",
       "1              1                                 -2   \n",
       "2              1                                 -1   \n",
       "3              1                                 -1   \n",
       "4              1                                  1   \n",
       "...          ...                                ...   \n",
       "14911          1                                  0   \n",
       "14912          1                                  0   \n",
       "14913        nan                                nan   \n",
       "14914        nan                                nan   \n",
       "14915          1                                  0   \n",
       "\n",
       "                                                      ID  \n",
       "0      SLC12A2ENST00000262461MCF7_BREASTACH-000019__-...  \n",
       "1      SLC12A2ENST00000262461GB1_CENTRAL_NERVOUS_SYST...  \n",
       "2      SLC12A2ENST00000262461SNGM_ENDOMETRIUMACH-0009...  \n",
       "3      SLC12A2ENST00000262461SKES1_BONEACH-000087__T_...  \n",
       "4      SLC12A2ENST00000262461MFE319_ENDOMETRIUMACH-00...  \n",
       "...                                                  ...  \n",
       "14911  PLA1AENST00000273371NCIH446_LUNGACH-000800__G_...  \n",
       "14912  ITGA8ENST00000378076NCIH650_LUNGACH-000945__C_...  \n",
       "14913  DLX5ENST00000222598HEC59_ENDOMETRIUMACH-000994...  \n",
       "14914  DLX5ENST00000222598HEC59_ENDOMETRIUMACH-000994...  \n",
       "14915  INPP5JENST00000331075HEC59_ENDOMETRIUMACH-0009...  \n",
       "\n",
       "[14916 rows x 18 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Table to check if in silico mutations have been done correctly\n",
    "ccle_mutcheck_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add mutated sequence to cptac_mut_df\n",
    "cptac_mut_df = pd.merge(cptac_mut_df,cptac_seq_mut_df,how='left')\n",
    "ccle_mut_df = pd.merge(ccle_mut_df,ccle_seq_mut_df,how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Translation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_code = {'GAA': 'E', 'CGA': 'R', 'GTG': 'V', 'TAA': '*', 'CGT': 'R', 'ATA': 'I', 'GAC': 'D', 'TCG': 'S', \n",
    "                  'GAT': 'D', 'ATG': 'M', 'CTG': 'L', 'CTA': 'L', 'TAC': 'Y', 'GGA': 'G', 'CGG': 'R', 'AGC': 'S', \n",
    "                  'TCT': 'S', 'TGA': '*', 'AAA': 'K', 'ACC': 'T', 'ACA': 'T', 'TGC': 'C', 'AAG': 'K', 'GTC': 'V', \n",
    "                  'TCC': 'S', 'ACT': 'T', 'AGA': 'R', 'CTT': 'L', 'GCC': 'A', 'GTA': 'V', 'TAG': '*', 'CAA': 'Q', \n",
    "                  'CAC': 'H', 'GCT': 'A', 'TTA': 'L', 'CAT': 'H', 'CGC': 'R', 'TTC': 'F', 'ATT': 'I', 'GGC': 'G', \n",
    "                  'CAG': 'Q', 'AAC': 'N', 'CCC': 'P', 'GTT': 'V', 'AGG': 'R', 'TGT': 'C', 'CCG': 'P', 'GGG': 'G', \n",
    "                  'ATC': 'I', 'TTT': 'F', 'AAT': 'N', 'TCA': 'S', 'GAG': 'E', 'CCA': 'P', 'GCA': 'A', 'TAT': 'Y', \n",
    "                  'GGT': 'G', 'TGG': 'W', 'GCG': 'A', 'CTC': 'L', 'TTG': 'L', 'CCT': 'P', 'ACG': 'T', 'AGT': 'S'}\n",
    "\n",
    "def translation(nt_seq_list):\n",
    "    '''Function to translate in silico a list of cdna sequences'''\n",
    "    prot_seq_list = []\n",
    "    for nt_seq in tqdm(nt_seq_list):\n",
    "        if (nt_seq[0] is np.nan)|(nt_seq[0] == 'Sequence unavailable')|(nt_seq[0] == 'No_ref_match')|(nt_seq[0] == 'Short_seq_wt'):\n",
    "            prot_seq_ids = [nt_seq[1],nt_seq[2],nt_seq[3],nt_seq[0]]\n",
    "            prot_seq_list.append(prot_seq_ids)\n",
    "        else:\n",
    "            nt_seq_triplets = [nt_seq[0][i:i+3] for i in range(0, len(nt_seq[0]), 3)]\n",
    "            prot_seq = ''\n",
    "\n",
    "            for triplet in nt_seq_triplets:\n",
    "                if len(triplet) == 3:\n",
    "                    if 'N' in triplet:\n",
    "                        prot_seq += 'X' \n",
    "                    else:\n",
    "                        prot_seq += genetic_code[triplet]\n",
    "                        if '*' in prot_seq:\n",
    "                            break\n",
    "                else:\n",
    "                    break\n",
    "            prot_seq_ids = [nt_seq[1],nt_seq[2],nt_seq[3],prot_seq]\n",
    "            prot_seq_list.append(prot_seq_ids)\n",
    "    return(prot_seq_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translation_mut_table (df):\n",
    "    '''Function to prepare a table with translated mutated sequences'''\n",
    "    seq_mut_list = df[['seq_mut','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    prot_seq_mut_list = translation(seq_mut_list)\n",
    "    prot_seq_mut_df = pd.DataFrame(prot_seq_mut_list, columns=['Feature','#Uploaded_variation','Location','prot_seq'])\n",
    "    prot_mut_df = pd.merge(df,prot_seq_mut_df,how='left')\n",
    "    return (prot_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d24ece6c014d43199d9c25c5f790726e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=16401.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e36284b210ae4006bbf4608970efc124",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=14916.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#Translate sequences\n",
    "cptac_prot_mut_df = translation_mut_table(cptac_mut_df)\n",
    "ccle_prot_mut_df = translation_mut_table(ccle_mut_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translation_wt_table (df,seq_df):\n",
    "    '''Function to prepare a table with translated wt sequences'''\n",
    "    #Prepare cptac_df table for the merge\n",
    "    wt_df = df[['gene','Feature','#Uploaded_variation','Location','protein_mutation','Protein_position','Amino_acids','sample']][df['Phenotype']!='WT']\n",
    "    wt_df.drop_duplicates(keep='first',inplace=True)\n",
    "\n",
    "    #Merge seq with cptac table for all proteins\n",
    "    seq_wt_df = seq_df[['ensembl_transcript_id','coding']]\n",
    "    seq_wt_df.drop_duplicates(keep='first',inplace=True)\n",
    "    seq_wt_df = seq_wt_df.rename(columns={'ensembl_transcript_id':'Feature','coding':'seq_wt'})\n",
    "\n",
    "    wt_df = pd.merge(wt_df, seq_wt_df, how='left')\n",
    "\n",
    "    #Translate wt sequences\n",
    "    seq_wt_list = wt_df[['seq_wt','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    prot_seq_wt_list = translation(seq_wt_list)\n",
    "    prot_wt_df = pd.DataFrame(prot_seq_wt_list, columns=['Feature','#Uploaded_variation','Location','prot_seq'])\n",
    "    return (prot_wt_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/msanchezg/miniconda3/envs/cptac/lib/python3.7/site-packages/ipykernel_launcher.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7c467a0cab514910b865e5dd728530ac",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=351296.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f0f3879d78224f9d83c73129538755b5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=161924.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#Create a table with translated sequences\n",
    "cptac_prot_wt_df = translation_wt_table (cptac_df,seq_df)\n",
    "ccle_prot_wt_df = translation_wt_table (ccle_df,seq_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Search for c-term degrons in the generated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_cdegron(prot_list, cdegron_motif):\n",
    "    \"\"\"This function finds all proteins containing c-terminal degrons (cdegrons)\"\"\"\n",
    "    cdegron_seq_list = [] \n",
    "    for mutation in prot_list:\n",
    "        seq = mutation[0]\n",
    "        splitted = re.split('\\*', str(seq))\n",
    "        seq2 = splitted[0]\n",
    "        motif_re = cdegron_motif +'$'\n",
    "        motif_re = motif_re.replace('X', '.')\n",
    "        find_cdegron_motif = re.findall(motif_re, seq2)\n",
    "        if find_cdegron_motif != []:\n",
    "            mutation_list = []\n",
    "            mutation_list.append(mutation[1])\n",
    "            mutation_list.append(mutation[2])\n",
    "            mutation_list.append(mutation[3])\n",
    "            mutation_list.append(seq)\n",
    "            mutation_list.append(True)\n",
    "            cdegron_seq_list.append(mutation_list)\n",
    "    cdegron_df = pd.DataFrame(cdegron_seq_list,columns=['Feature','#Uploaded_variation','Location','prot_seq',cdegron_motif])\n",
    "    return cdegron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "cdegron_motifs = ['GG', 'RG', 'PG', 'XR', 'RXXG', 'EE', 'RXX', 'VX', 'AX', 'A']\n",
    "\n",
    "def find_cdegron_list(prot_list, cdegron_motifs):\n",
    "    \"\"\"This function finds all cdegron motifs provided in a list\"\"\"\n",
    "    col_names=['Feature','#Uploaded_variation','Location','prot_seq']\n",
    "    all_cdegron_df = pd.DataFrame(columns=col_names)\n",
    "    for motif in cdegron_motifs:\n",
    "        cdegron_df = find_cdegron(prot_list, motif)\n",
    "        all_cdegron_df = pd.concat([all_cdegron_df,cdegron_df])\n",
    "    all_cdegron_df = all_cdegron_df.fillna(False)\n",
    "    dupl = all_cdegron_df[all_cdegron_df[['Feature','#Uploaded_variation','Location']].duplicated()].values.tolist()\n",
    "    df2 = pd.DataFrame()\n",
    "    for seq in dupl:\n",
    "        df1 = all_cdegron_df[(all_cdegron_df['Feature']==seq[0])&(all_cdegron_df['#Uploaded_variation']==seq[1])&(all_cdegron_df['Location']==seq[2])].groupby(['Feature','#Uploaded_variation','Location','prot_seq'], as_index=False).sum()\n",
    "        df2 = pd.concat([df2,df1])\n",
    "    all_cdegron_df.drop_duplicates(subset=['Feature','#Uploaded_variation','Location','prot_seq'], keep=False, inplace=True)\n",
    "    df2.drop_duplicates(keep='first', inplace=True)\n",
    "    all_cdegron_df = pd.concat([all_cdegron_df,df2],ignore_index=True)\n",
    "    return all_cdegron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cdegron_table (prot_mut_df, prot_wt_df, cdegron_motifs, df,seq_df):\n",
    "    '''This function creates a table with c-degrons annotated per sequence'''\n",
    "    \n",
    "    #Search cdegron motifs in cptac_mut_df\n",
    "    print('Search cdegron motifs in cptac_mut_df')\n",
    "    \n",
    "    prot_seq_mut_list = prot_mut_df[['prot_seq','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    cdegron_mut_df = find_cdegron_list(prot_seq_mut_list,cdegron_motifs)\n",
    "\n",
    "    #Add proteins without cterm degrons to the table\n",
    "    print('Add proteins without cterm degrons to the table')\n",
    "    \n",
    "    prot_seq_mut_df = prot_mut_df[['gene','Feature','#Uploaded_variation','Location','prot_seq']]\n",
    "    cdegron_mut_df = pd.merge(prot_seq_mut_df,cdegron_mut_df,how='left')\n",
    "    cdegron_mut_df = cdegron_mut_df.fillna(False)\n",
    "    \n",
    "    #Find wt proteins with cterm degrons\n",
    "    print('Find wt proteins with cterm degrons')\n",
    "    \n",
    "    prot_wt_list = prot_wt_df[['prot_seq','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    cdegron_wt_df = find_cdegron_list(prot_wt_list,cdegron_motifs)\n",
    "\n",
    "    #Add proteins without cterm degrons to the table\n",
    "    print('Add proteins without cterm degrons to the table')\n",
    "    \n",
    "    cdegron_wt_df = pd.merge(prot_wt_df,cdegron_wt_df,how='left')\n",
    "    cdegron_wt_df = cdegron_wt_df.fillna(False)\n",
    "\n",
    "    #Add 'gene' and 'Phenotype' to cdegron_wt_df\n",
    "    print('Add gene and Phenotype to cdegron_wt_df')\n",
    "    \n",
    "    id_df = df[['gene','Feature']][df['Phenotype']!='WT']\n",
    "    id_df.drop_duplicates(keep='first',inplace=True)\n",
    "    cdegron_wt_df = pd.merge(cdegron_wt_df,id_df,how='left')\n",
    "\n",
    "    cdegron_wt2_df = cdegron_wt_df.drop(columns=['Feature','#Uploaded_variation','Location'])\n",
    "    phenotype_wt = 'WT'\n",
    "    cdegron_wt2_df['Phenotype'] = phenotype_wt\n",
    "\n",
    "    #Add phenotype to cdegron_mut_df\n",
    "    print('Add phenotype to cdegron_mut_df')\n",
    "    \n",
    "    phenotype_mut = df[['Feature','#Uploaded_variation','Location','Phenotype']][(df['Phenotype']=='stop_gained')|(df['Phenotype']=='frameshift_variant')]\n",
    "    cdegron_mut_df = pd.merge(cdegron_mut_df,phenotype_mut,how='left')\n",
    "    cdegron_mut_df = pd.merge(cdegron_mut_df,id_df,how='left')\n",
    "\n",
    "    #Concat wt and mut\n",
    "    print('Concat wt and mut')\n",
    "    \n",
    "    cdegron_df = pd.concat([cdegron_wt2_df,cdegron_mut_df])\n",
    "    cdegron_df.drop_duplicates(keep='first',inplace=True) \n",
    "    return (cdegron_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPTAC dataset\n",
      "Search cdegron motifs in cptac_mut_df\n",
      "Add proteins without cterm degrons to the table\n",
      "Find wt proteins with cterm degrons\n",
      "Add proteins without cterm degrons to the table\n",
      "Add gene and Phenotype to cdegron_wt_df\n",
      "Add phenotype to cdegron_mut_df\n",
      "Concat wt and mut\n",
      "\n",
      "CCLE dataset\n",
      "Search cdegron motifs in cptac_mut_df\n",
      "Add proteins without cterm degrons to the table\n",
      "Find wt proteins with cterm degrons\n",
      "Add proteins without cterm degrons to the table\n",
      "Add gene and Phenotype to cdegron_wt_df\n",
      "Add phenotype to cdegron_mut_df\n",
      "Concat wt and mut\n"
     ]
    }
   ],
   "source": [
    "#Create tables with c-degron annotations\n",
    "print('CPTAC dataset')\n",
    "cptac_cdegron_df = cdegron_table (cptac_prot_mut_df, cptac_prot_wt_df, cdegron_motifs, cptac_df, seq_df)\n",
    "print('\\nCCLE dataset')\n",
    "ccle_cdegron_df = cdegron_table (ccle_prot_mut_df, ccle_prot_wt_df, cdegron_motifs, ccle_df, seq_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_seq</th>\n",
       "      <th>GG</th>\n",
       "      <th>RG</th>\n",
       "      <th>PG</th>\n",
       "      <th>XR</th>\n",
       "      <th>RXXG</th>\n",
       "      <th>EE</th>\n",
       "      <th>RXX</th>\n",
       "      <th>VX</th>\n",
       "      <th>AX</th>\n",
       "      <th>A</th>\n",
       "      <th>gene</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Feature</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGKNKLLHPSLVLLLLVLLPTDASVSGKPQYMVLVPSLLHTETTEK...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>A2M</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MWAQLLLGMLALSPAIAEELPNYLVTLPARLNFPSVQKVCLDLSPG...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MCSLGLFPPPPPRGQVTLYEHNNELVTGSSYESPPPDFRGQWINLP...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MSKEERPGREEILECQVMWEPDSKKNTQMDRFRAAVGAACGLALES...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AACS</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MAAPAPVTRQVSGAAALVPAPSGPDSGQPLAAAVAELPVLDARGQR...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>AAED1</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16396</th>\n",
       "      <td>MTMYLWLKLLAFGFAFLDTEVFVTGQSPTPSPTGLTTAKMPSVPLS...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PTPRC</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000442510</td>\n",
       "      <td>C3N-01825__-__AACTCCT</td>\n",
       "      <td>chr1:198748182-198748183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16397</th>\n",
       "      <td>MALRELKVCLLGDTGVGKSSIVWRFVEDSFDPNINPTIGASFMTKT...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>RAB22A</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000244040</td>\n",
       "      <td>C3N-01825__-__AAAAAAA</td>\n",
       "      <td>chr20:58354224-58354225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16398</th>\n",
       "      <td>MKIETSGYNSDKLICRGFIGTPAPPVCDSKFLLSPSSDVRMIPCRA...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>SLC7A2</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000004531</td>\n",
       "      <td>C3N-01825__-__ATTCA</td>\n",
       "      <td>chr8:17548817-17548818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16399</th>\n",
       "      <td>MPSTNRAGSLKDPEIAELFFKEDPEKLFTDLREIGHGSFGAVYFAR...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TAOK1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000261716</td>\n",
       "      <td>C3N-01825__-__TTCATTG</td>\n",
       "      <td>chr17:29510944-29510945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16400</th>\n",
       "      <td>MAEPRRVAFISLSPVRRREAEYPGPEREPEYPREPPRLEPQPYREP...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>UBN2</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000473989</td>\n",
       "      <td>C3N-01825__-__TGGGTCT</td>\n",
       "      <td>chr7:139237094-139237095</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29247 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                prot_seq     GG     RG     PG  \\\n",
       "0      MGKNKLLHPSLVLLLLVLLPTDASVSGKPQYMVLVPSLLHTETTEK...  False  False  False   \n",
       "5      MWAQLLLGMLALSPAIAEELPNYLVTLPARLNFPSVQKVCLDLSPG...  False  False  False   \n",
       "10     MCSLGLFPPPPPRGQVTLYEHNNELVTGSSYESPPPDFRGQWINLP...  False  False  False   \n",
       "12     MSKEERPGREEILECQVMWEPDSKKNTQMDRFRAAVGAACGLALES...  False  False  False   \n",
       "15     MAAPAPVTRQVSGAAALVPAPSGPDSGQPLAAAVAELPVLDARGQR...  False  False  False   \n",
       "...                                                  ...    ...    ...    ...   \n",
       "16396  MTMYLWLKLLAFGFAFLDTEVFVTGQSPTPSPTGLTTAKMPSVPLS...  False  False  False   \n",
       "16397  MALRELKVCLLGDTGVGKSSIVWRFVEDSFDPNINPTIGASFMTKT...  False  False  False   \n",
       "16398  MKIETSGYNSDKLICRGFIGTPAPPVCDSKFLLSPSSDVRMIPCRA...  False  False  False   \n",
       "16399  MPSTNRAGSLKDPEIAELFFKEDPEKLFTDLREIGHGSFGAVYFAR...  False  False  False   \n",
       "16400  MAEPRRVAFISLSPVRRREAEYPGPEREPEYPREPPRLEPQPYREP...  False  False  False   \n",
       "\n",
       "          XR   RXXG     EE    RXX     VX     AX      A    gene  \\\n",
       "0      False  False  False  False  False  False   True     A2M   \n",
       "5      False  False  False  False  False  False  False   A2ML1   \n",
       "10     False  False  False  False  False  False  False    AAAS   \n",
       "12     False  False  False  False  False  False  False    AACS   \n",
       "15     False  False  False  False  False  False  False   AAED1   \n",
       "...      ...    ...    ...    ...    ...    ...    ...     ...   \n",
       "16396  False  False  False  False  False  False  False   PTPRC   \n",
       "16397  False  False  False  False  False  False  False  RAB22A   \n",
       "16398  False  False  False  False  False  False  False  SLC7A2   \n",
       "16399  False  False  False  False  False  False  False   TAOK1   \n",
       "16400  False  False  False  False  False  False  False    UBN2   \n",
       "\n",
       "                Phenotype          Feature    #Uploaded_variation  \\\n",
       "0                      WT              NaN                    NaN   \n",
       "5                      WT              NaN                    NaN   \n",
       "10                     WT              NaN                    NaN   \n",
       "12                     WT              NaN                    NaN   \n",
       "15                     WT              NaN                    NaN   \n",
       "...                   ...              ...                    ...   \n",
       "16396  frameshift_variant  ENST00000442510  C3N-01825__-__AACTCCT   \n",
       "16397  frameshift_variant  ENST00000244040  C3N-01825__-__AAAAAAA   \n",
       "16398  frameshift_variant  ENST00000004531    C3N-01825__-__ATTCA   \n",
       "16399  frameshift_variant  ENST00000261716  C3N-01825__-__TTCATTG   \n",
       "16400  frameshift_variant  ENST00000473989  C3N-01825__-__TGGGTCT   \n",
       "\n",
       "                       Location  \n",
       "0                           NaN  \n",
       "5                           NaN  \n",
       "10                          NaN  \n",
       "12                          NaN  \n",
       "15                          NaN  \n",
       "...                         ...  \n",
       "16396  chr1:198748182-198748183  \n",
       "16397   chr20:58354224-58354225  \n",
       "16398    chr8:17548817-17548818  \n",
       "16399   chr17:29510944-29510945  \n",
       "16400  chr7:139237094-139237095  \n",
       "\n",
       "[29247 rows x 16 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_cdegron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_seq</th>\n",
       "      <th>GG</th>\n",
       "      <th>RG</th>\n",
       "      <th>PG</th>\n",
       "      <th>XR</th>\n",
       "      <th>RXXG</th>\n",
       "      <th>EE</th>\n",
       "      <th>RXX</th>\n",
       "      <th>VX</th>\n",
       "      <th>AX</th>\n",
       "      <th>A</th>\n",
       "      <th>gene</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Feature</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEPRPTAPSSGAPGLAGVGETPSAAALAAARVELPGTAVPSVPEDA...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>SLC12A2</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>MLSGKKAAAAAAAAAAAATGTEAGPGTAGGSENGSEVAAQPAGLSG...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MYNMMETELKPPGPQQTSGGGGGNSTAAAAGGNQKNSPDRVKRPMN...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MDSAAAAFALDKPALGPGPPPPPPALGPGDCAQARKNFSVSHLLDL...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PRRX2</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MVLLESEQFLTELTRLFQKCRTSGSVYITLKKYDGRTKPIPKKGTV...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP14</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14911</th>\n",
       "      <td>MPPGPWESCFWVGGLILWLSVGSSGDAPPTPQPKCADFQSANLFEG...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>PLA1A</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000273371</td>\n",
       "      <td>ACH-000800__G__T</td>\n",
       "      <td>chr3:119608918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14912</th>\n",
       "      <td>MSPGASRGPRGSQAPLIAPLCCAAAALGMLLWSPACQAFNLDVEKL...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ITGA8</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000378076</td>\n",
       "      <td>ACH-000945__C__A</td>\n",
       "      <td>chr10:15688006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14913</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000222598</td>\n",
       "      <td>ACH-000994__AA__-</td>\n",
       "      <td>chr7:97022295-97022296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14914</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>DLX5</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000222598</td>\n",
       "      <td>ACH-000994__C__-</td>\n",
       "      <td>chr7:97024466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14915</th>\n",
       "      <td>MEGQSSRGSRRPGTRAGLGSLPMPQGVAQTGAPSKVDSSFQLPAKK...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>INPP5J</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000331075</td>\n",
       "      <td>ACH-000994__C__T</td>\n",
       "      <td>chr22:31125775</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24325 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                prot_seq     GG     RG     PG  \\\n",
       "0      MEPRPTAPSSGAPGLAGVGETPSAAALAAARVELPGTAVPSVPEDA...  False  False  False   \n",
       "32     MLSGKKAAAAAAAAAAAATGTEAGPGTAGGSENGSEVAAQPAGLSG...  False  False  False   \n",
       "46     MYNMMETELKPPGPQQTSGGGGGNSTAAAAGGNQKNSPDRVKRPMN...  False  False  False   \n",
       "58     MDSAAAAFALDKPALGPGPPPPPPALGPGDCAQARKNFSVSHLLDL...  False  False  False   \n",
       "59     MVLLESEQFLTELTRLFQKCRTSGSVYITLKKYDGRTKPIPKKGTV...  False  False  False   \n",
       "...                                                  ...    ...    ...    ...   \n",
       "14911  MPPGPWESCFWVGGLILWLSVGSSGDAPPTPQPKCADFQSANLFEG...  False  False  False   \n",
       "14912  MSPGASRGPRGSQAPLIAPLCCAAAALGMLLWSPACQAFNLDVEKL...  False  False  False   \n",
       "14913                                              False  False  False  False   \n",
       "14914                                              False  False  False  False   \n",
       "14915  MEGQSSRGSRRPGTRAGLGSLPMPQGVAQTGAPSKVDSSFQLPAKK...  False  False  False   \n",
       "\n",
       "          XR   RXXG     EE    RXX     VX     AX      A     gene  \\\n",
       "0      False  False  False  False  False  False  False  SLC12A2   \n",
       "32     False  False  False  False  False  False  False    KDM1A   \n",
       "46     False  False  False  False  False  False  False     SOX2   \n",
       "58     False  False  False  False   True  False  False    PRRX2   \n",
       "59     False  False  False  False  False   True  False    SRP14   \n",
       "...      ...    ...    ...    ...    ...    ...    ...      ...   \n",
       "14911  False  False  False  False  False  False  False    PLA1A   \n",
       "14912  False  False  False  False  False  False  False    ITGA8   \n",
       "14913  False  False  False  False  False  False  False     DLX5   \n",
       "14914  False  False  False  False  False  False  False     DLX5   \n",
       "14915  False  False  False  False  False  False  False   INPP5J   \n",
       "\n",
       "                Phenotype          Feature #Uploaded_variation  \\\n",
       "0                      WT              NaN                 NaN   \n",
       "32                     WT              NaN                 NaN   \n",
       "46                     WT              NaN                 NaN   \n",
       "58                     WT              NaN                 NaN   \n",
       "59                     WT              NaN                 NaN   \n",
       "...                   ...              ...                 ...   \n",
       "14911         stop_gained  ENST00000273371    ACH-000800__G__T   \n",
       "14912         stop_gained  ENST00000378076    ACH-000945__C__A   \n",
       "14913  frameshift_variant  ENST00000222598   ACH-000994__AA__-   \n",
       "14914  frameshift_variant  ENST00000222598    ACH-000994__C__-   \n",
       "14915         stop_gained  ENST00000331075    ACH-000994__C__T   \n",
       "\n",
       "                     Location  \n",
       "0                         NaN  \n",
       "32                        NaN  \n",
       "46                        NaN  \n",
       "58                        NaN  \n",
       "59                        NaN  \n",
       "...                       ...  \n",
       "14911          chr3:119608918  \n",
       "14912          chr10:15688006  \n",
       "14913  chr7:97022295-97022296  \n",
       "14914           chr7:97024466  \n",
       "14915          chr22:31125775  \n",
       "\n",
       "[24325 rows x 16 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_cdegron_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save table with cdegron annotation\n",
    "cptac_cdegron_df.to_csv(r'cdegron_wtnsfs_annot_cptac.tsv', header = True, index = None, sep = '\\t')\n",
    "ccle_cdegron_df.to_csv(r'cdegron_wtnsfs_annot_ccle.tsv', header = True, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Get c-terminal amino acid (last aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_last_aa(prot_list, aa):\n",
    "    \"\"\"This function finds all proteins containing c-terminal degrons (cdegrons)\"\"\"\n",
    "    aa_seq_list = [] \n",
    "    for mutation in prot_list:\n",
    "        seq = mutation[0]\n",
    "        splitted = re.split('\\*', str(seq))\n",
    "        seq2 = splitted[0]\n",
    "        aa_re = aa +'$'\n",
    "        find_aa = re.findall(aa_re, seq2)\n",
    "        if find_aa != []:\n",
    "            mutation_list = []\n",
    "            mutation_list.append(mutation[1])\n",
    "            mutation_list.append(mutation[2])\n",
    "            mutation_list.append(mutation[3])\n",
    "            mutation_list.append(seq)\n",
    "            mutation_list.append(aa)\n",
    "            aa_seq_list.append(mutation_list)\n",
    "    aa_df = pd.DataFrame(aa_seq_list,columns=['Feature','#Uploaded_variation','Location','prot_seq','last_aa'])\n",
    "    return aa_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_list = ['A','C','D','E','F','G','H','I','K','Q','L','M','N','P','R','S','T','V','W','Y']\n",
    "\n",
    "def get_last_aa_list(prot_list, aa_list):\n",
    "    \"\"\"This function finds all cdegron motifs provided in a list\"\"\"\n",
    "    col_names=['Feature','#Uploaded_variation','Location','prot_seq']\n",
    "    all_aa_df = pd.DataFrame()\n",
    "    for aa in aa_list:\n",
    "        aa_df = get_last_aa(prot_list, aa)\n",
    "        all_aa_df = pd.concat([all_aa_df,aa_df],ignore_index=True)\n",
    "    return all_aa_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def last_aa_table (prot_mut_df,aa_list, prot_wt_df, df):\n",
    "    '''This function annotates all c-terminal amino acids (last aa) from mutated and wt sequences'''\n",
    "    \n",
    "    prot_seq_mut_list = prot_mut_df[['prot_seq','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    aa_mut_df = get_last_aa_list(prot_seq_mut_list,aa_list)\n",
    "\n",
    "    #Get last aa from wt sequences\n",
    "    prot_wt_list = prot_wt_df[['prot_seq','Feature','#Uploaded_variation','Location']].values.tolist()\n",
    "    aa_wt_df = get_last_aa_list(prot_wt_list,aa_list)\n",
    "\n",
    "    #Add 'gene' and 'Phenotype'\n",
    "    id_df = cptac_df[['gene','Feature']][cptac_df['Phenotype']!='WT']\n",
    "    id_df.drop_duplicates(keep='first',inplace=True)\n",
    "    aa_wt_df = pd.merge(aa_wt_df,id_df,how='left')\n",
    "\n",
    "    aa_wt2_df = aa_wt_df.drop(columns=['Feature','#Uploaded_variation','Location'])\n",
    "    phenotype_wt = 'WT'\n",
    "    aa_wt2_df['Phenotype'] = phenotype_wt\n",
    "\n",
    "    phenotype_mut = df[['Feature','#Uploaded_variation','Location','Phenotype']][(df['Phenotype']=='stop_gained')|(df['Phenotype']=='frameshift_variant')]\n",
    "    aa_mut_df = pd.merge(aa_mut_df,phenotype_mut,how='left')\n",
    "    aa_mut_df = pd.merge(aa_mut_df,id_df,how='left')\n",
    "\n",
    "    #concat wt and mut\n",
    "    aa_df = pd.concat([aa_wt2_df,aa_mut_df])\n",
    "    aa_df.drop_duplicates(keep='first',inplace=True) \n",
    "    return (aa_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create tables with last aa annotation\n",
    "cptac_aa_df = last_aa_table (cptac_prot_mut_df, aa_list, cptac_prot_wt_df, cptac_df)\n",
    "ccle_aa_df = last_aa_table (ccle_prot_mut_df, aa_list, ccle_prot_wt_df, ccle_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_seq</th>\n",
       "      <th>last_aa</th>\n",
       "      <th>gene</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Feature</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGKNKLLHPSLVLLLLVLLPTDASVSGKPQYMVLVPSLLHTETTEK...</td>\n",
       "      <td>A</td>\n",
       "      <td>A2M</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MRGPPAWPLRLLEPPSPAEPGRLLPVACVWAAASRVPGSLSPFTGL...</td>\n",
       "      <td>A</td>\n",
       "      <td>ABCB10</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MDALCGSGELGSKFWDSNLSVHTENPDLTPCFQNSLLAWVPCIYLW...</td>\n",
       "      <td>A</td>\n",
       "      <td>ABCC3</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MALLRGVFVVAAKRTPFGAYGGLLKDFTATDLSEFAAKAALSAGKV...</td>\n",
       "      <td>A</td>\n",
       "      <td>ACAA2</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MASSFLPAGAITGDSGGELSSGDDSGEVEFPHSPEIEETSCLAELF...</td>\n",
       "      <td>A</td>\n",
       "      <td>ACBD6</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16068</th>\n",
       "      <td>MIKSQESLTLEDVAVEFTWEEWQLLGPAQKDLY*</td>\n",
       "      <td>Y</td>\n",
       "      <td>ZNF613</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000293471</td>\n",
       "      <td>C3N-01520__C__T</td>\n",
       "      <td>chr19:51940293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16069</th>\n",
       "      <td>MDLPALLPAPTARGGQHGGGPGPLRRAPAPLGASPARRRLLLVRGP...</td>\n",
       "      <td>Y</td>\n",
       "      <td>ZXDC</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000389709</td>\n",
       "      <td>C3L-00771__G__A</td>\n",
       "      <td>chr3:126459707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16070</th>\n",
       "      <td>MATRSSRRESRLPFLFTLVALLPPGALCEVWTQRLHGGSAPLPQDR...</td>\n",
       "      <td>Y</td>\n",
       "      <td>SORL1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000260197</td>\n",
       "      <td>C3L-00908__-__AAAAAAAA</td>\n",
       "      <td>chr11:121605195-121605196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16071</th>\n",
       "      <td>MAGEVSAATGRFSLERLGLPGLALAAALLLLALCLLVRRTRRPGEP...</td>\n",
       "      <td>Y</td>\n",
       "      <td>CYP7B1</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000310193</td>\n",
       "      <td>C3N-00545__C__A</td>\n",
       "      <td>chr8:64615949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16072</th>\n",
       "      <td>MATAATEEPFPFHGLLPKKETGAASFLCRYPEYDGRGVLIAVLDTG...</td>\n",
       "      <td>Y</td>\n",
       "      <td>TPP2</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>ENST00000376052</td>\n",
       "      <td>C3N-00545__-__TTAGATTAATATGCTT</td>\n",
       "      <td>chr13:102657135-102657136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28771 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                prot_seq last_aa    gene  \\\n",
       "0      MGKNKLLHPSLVLLLLVLLPTDASVSGKPQYMVLVPSLLHTETTEK...       A     A2M   \n",
       "5      MRGPPAWPLRLLEPPSPAEPGRLLPVACVWAAASRVPGSLSPFTGL...       A  ABCB10   \n",
       "8      MDALCGSGELGSKFWDSNLSVHTENPDLTPCFQNSLLAWVPCIYLW...       A   ABCC3   \n",
       "9      MALLRGVFVVAAKRTPFGAYGGLLKDFTATDLSEFAAKAALSAGKV...       A   ACAA2   \n",
       "11     MASSFLPAGAITGDSGGELSSGDDSGEVEFPHSPEIEETSCLAELF...       A   ACBD6   \n",
       "...                                                  ...     ...     ...   \n",
       "16068                 MIKSQESLTLEDVAVEFTWEEWQLLGPAQKDLY*       Y  ZNF613   \n",
       "16069  MDLPALLPAPTARGGQHGGGPGPLRRAPAPLGASPARRRLLLVRGP...       Y    ZXDC   \n",
       "16070  MATRSSRRESRLPFLFTLVALLPPGALCEVWTQRLHGGSAPLPQDR...       Y   SORL1   \n",
       "16071  MAGEVSAATGRFSLERLGLPGLALAAALLLLALCLLVRRTRRPGEP...       Y  CYP7B1   \n",
       "16072  MATAATEEPFPFHGLLPKKETGAASFLCRYPEYDGRGVLIAVLDTG...       Y    TPP2   \n",
       "\n",
       "                Phenotype          Feature             #Uploaded_variation  \\\n",
       "0                      WT              NaN                             NaN   \n",
       "5                      WT              NaN                             NaN   \n",
       "8                      WT              NaN                             NaN   \n",
       "9                      WT              NaN                             NaN   \n",
       "11                     WT              NaN                             NaN   \n",
       "...                   ...              ...                             ...   \n",
       "16068         stop_gained  ENST00000293471                 C3N-01520__C__T   \n",
       "16069         stop_gained  ENST00000389709                 C3L-00771__G__A   \n",
       "16070  frameshift_variant  ENST00000260197          C3L-00908__-__AAAAAAAA   \n",
       "16071         stop_gained  ENST00000310193                 C3N-00545__C__A   \n",
       "16072         stop_gained  ENST00000376052  C3N-00545__-__TTAGATTAATATGCTT   \n",
       "\n",
       "                        Location  \n",
       "0                            NaN  \n",
       "5                            NaN  \n",
       "8                            NaN  \n",
       "9                            NaN  \n",
       "11                           NaN  \n",
       "...                          ...  \n",
       "16068             chr19:51940293  \n",
       "16069             chr3:126459707  \n",
       "16070  chr11:121605195-121605196  \n",
       "16071              chr8:64615949  \n",
       "16072  chr13:102657135-102657136  \n",
       "\n",
       "[28771 rows x 7 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_aa_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_seq</th>\n",
       "      <th>last_aa</th>\n",
       "      <th>gene</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Feature</th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEGSGGGAGERAPLLGARRAAAAAAAAGAFAGRRAACGAVLLTELL...</td>\n",
       "      <td>A</td>\n",
       "      <td>SLC15A4</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MACARPLISVYSEKGESSGKNVTLPAVFKAPIRPDIVNFVHTNLRK...</td>\n",
       "      <td>A</td>\n",
       "      <td>RPL4</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MAATASAGVPATVSEKQEFYQLLKNLINPSCMVRRQAEEIYENIPG...</td>\n",
       "      <td>A</td>\n",
       "      <td>RANBP6</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MKNQDKKNGAAKQSNPKSSPGQPEAGPEGAQERPSQAAPAVEAEGP...</td>\n",
       "      <td>A</td>\n",
       "      <td>TXLNA</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>MGDAADPREMRKTFIVPAIKPFDHYDFSRAKIACNLAWLVAKAFGT...</td>\n",
       "      <td>A</td>\n",
       "      <td>CAMSAP2</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14650</th>\n",
       "      <td>MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...</td>\n",
       "      <td>Y</td>\n",
       "      <td>MS4A1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000534668</td>\n",
       "      <td>ACH-000817__-__A</td>\n",
       "      <td>chr11:60468274-60468275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14651</th>\n",
       "      <td>MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...</td>\n",
       "      <td>Y</td>\n",
       "      <td>MS4A1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000534668</td>\n",
       "      <td>ACH-000388__-__A</td>\n",
       "      <td>chr11:60468274-60468275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14652</th>\n",
       "      <td>MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...</td>\n",
       "      <td>Y</td>\n",
       "      <td>MS4A1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000534668</td>\n",
       "      <td>ACH-000151__-__A</td>\n",
       "      <td>chr11:60468274-60468275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14653</th>\n",
       "      <td>MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...</td>\n",
       "      <td>Y</td>\n",
       "      <td>MS4A1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000534668</td>\n",
       "      <td>ACH-000611__-__A</td>\n",
       "      <td>chr11:60468274-60468275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14654</th>\n",
       "      <td>MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...</td>\n",
       "      <td>Y</td>\n",
       "      <td>MS4A1</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>ENST00000534668</td>\n",
       "      <td>ACH-000365__-__A</td>\n",
       "      <td>chr11:60468274-60468275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23944 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                prot_seq last_aa     gene  \\\n",
       "0      MEGSGGGAGERAPLLGARRAAAAAAAAGAFAGRRAACGAVLLTELL...       A  SLC15A4   \n",
       "19     MACARPLISVYSEKGESSGKNVTLPAVFKAPIRPDIVNFVHTNLRK...       A     RPL4   \n",
       "37     MAATASAGVPATVSEKQEFYQLLKNLINPSCMVRRQAEEIYENIPG...       A   RANBP6   \n",
       "60     MKNQDKKNGAAKQSNPKSSPGQPEAGPEGAQERPSQAAPAVEAEGP...       A    TXLNA   \n",
       "81     MGDAADPREMRKTFIVPAIKPFDHYDFSRAKIACNLAWLVAKAFGT...       A  CAMSAP2   \n",
       "...                                                  ...     ...      ...   \n",
       "14650  MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...       Y    MS4A1   \n",
       "14651  MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...       Y    MS4A1   \n",
       "14652  MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...       Y    MS4A1   \n",
       "14653  MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...       Y    MS4A1   \n",
       "14654  MTTPRNSVNGTFPAEPMKGPIAMQSGPKPLFRRMSSLVGPTQSFFM...       Y    MS4A1   \n",
       "\n",
       "                Phenotype          Feature #Uploaded_variation  \\\n",
       "0                      WT              NaN                 NaN   \n",
       "19                     WT              NaN                 NaN   \n",
       "37                     WT              NaN                 NaN   \n",
       "60                     WT              NaN                 NaN   \n",
       "81                     WT              NaN                 NaN   \n",
       "...                   ...              ...                 ...   \n",
       "14650  frameshift_variant  ENST00000534668    ACH-000817__-__A   \n",
       "14651  frameshift_variant  ENST00000534668    ACH-000388__-__A   \n",
       "14652  frameshift_variant  ENST00000534668    ACH-000151__-__A   \n",
       "14653  frameshift_variant  ENST00000534668    ACH-000611__-__A   \n",
       "14654  frameshift_variant  ENST00000534668    ACH-000365__-__A   \n",
       "\n",
       "                      Location  \n",
       "0                          NaN  \n",
       "19                         NaN  \n",
       "37                         NaN  \n",
       "60                         NaN  \n",
       "81                         NaN  \n",
       "...                        ...  \n",
       "14650  chr11:60468274-60468275  \n",
       "14651  chr11:60468274-60468275  \n",
       "14652  chr11:60468274-60468275  \n",
       "14653  chr11:60468274-60468275  \n",
       "14654  chr11:60468274-60468275  \n",
       "\n",
       "[23944 rows x 7 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_aa_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save table with last aa annotation\n",
    "cptac_aa_df.to_csv(r'last_aa_annot_cptac.tsv', header = True, index = None, sep = '\\t')\n",
    "ccle_aa_df.to_csv(r'last_aa_annot_ccle.tsv', header = True, index = None, sep = '\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:cptac]",
   "language": "python",
   "name": "conda-env-cptac-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
