{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NMD score tables pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from tqdm.notebook import tqdm\n",
    "from operator import add"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare input nmd_scores table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/msanchezg/miniconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (5,8) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>nmd_score</th>\n",
       "      <th>alignID.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69091</td>\n",
       "      <td>70005</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>uc001aal.1</td>\n",
       "      <td>uc001aal.1</td>\n",
       "      <td>1</td>\n",
       "      <td>uc001aal.1.1</td>\n",
       "      <td>-0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0....</td>\n",
       "      <td>uc001aal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>182709</td>\n",
       "      <td>182746</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>uc057auo.1</td>\n",
       "      <td>uc057auo.1</td>\n",
       "      <td>1</td>\n",
       "      <td>uc057auo.1.1</td>\n",
       "      <td>0.23,0.23,0.23,0.24,0.24,0.24,0.24,0.24,0.24,...</td>\n",
       "      <td>uc057auo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>183114</td>\n",
       "      <td>183240</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>uc057auo.1</td>\n",
       "      <td>uc057auo.1</td>\n",
       "      <td>2</td>\n",
       "      <td>uc057auo.1.2</td>\n",
       "      <td>0.31,0.31,0.31,0.31,0.32,0.32,0.32,0.32,0.32,...</td>\n",
       "      <td>uc057auo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>183922</td>\n",
       "      <td>184155</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>uc057auo.1</td>\n",
       "      <td>uc057auo.1</td>\n",
       "      <td>3</td>\n",
       "      <td>uc057auo.1.3</td>\n",
       "      <td>0.17,0.17,0.17,0.17,0.18,0.17,0.17,0.18,0.18,...</td>\n",
       "      <td>uc057auo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>185220</td>\n",
       "      <td>185350</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>uc057aup.1</td>\n",
       "      <td>uc057aup.1</td>\n",
       "      <td>1</td>\n",
       "      <td>uc057aup.1.1</td>\n",
       "      <td>0.1,0.1,0.1,0.07,0.08,0.08,0.08,0.08,0.09,0.1...</td>\n",
       "      <td>uc057aup</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702477</th>\n",
       "      <td>chrY</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57190204</td>\n",
       "      <td>57190375</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>5</td>\n",
       "      <td>uc004fxm.2.5</td>\n",
       "      <td>0.58,0.58,0.58,0.61,0.61,0.62,0.61,0.61,0.61,...</td>\n",
       "      <td>uc004fxm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702478</th>\n",
       "      <td>chrY</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57190969</td>\n",
       "      <td>57191085</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>6</td>\n",
       "      <td>uc004fxm.2.6</td>\n",
       "      <td>0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,...</td>\n",
       "      <td>uc004fxm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702479</th>\n",
       "      <td>chrY</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57191798</td>\n",
       "      <td>57191999</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>7</td>\n",
       "      <td>uc004fxm.2.7</td>\n",
       "      <td>0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.57,...</td>\n",
       "      <td>uc004fxm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702480</th>\n",
       "      <td>chrY</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57192603</td>\n",
       "      <td>57192708</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>8</td>\n",
       "      <td>uc004fxm.2.8</td>\n",
       "      <td>0.6,0.65,0.66,0.64,0.59,0.56,0.56,0.54,0.54,0...</td>\n",
       "      <td>uc004fxm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702481</th>\n",
       "      <td>chrY</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57196336</td>\n",
       "      <td>57196369</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>uc004fxm.2</td>\n",
       "      <td>9</td>\n",
       "      <td>uc004fxm.2.9</td>\n",
       "      <td>0.05,0.06,0.06,0.06,0.06,0.06,0.06,0.06,0.06,...</td>\n",
       "      <td>uc004fxm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>702482 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM Feature     Start       End Strand Frame     gene_id  \\\n",
       "0       chr1     CDS     69091     70005      +     0  uc001aal.1   \n",
       "1       chr1     CDS    182709    182746      +     0  uc057auo.1   \n",
       "2       chr1     CDS    183114    183240      +     1  uc057auo.1   \n",
       "3       chr1     CDS    183922    184155      +     0  uc057auo.1   \n",
       "4       chr1     CDS    185220    185350      -     2  uc057aup.1   \n",
       "...      ...     ...       ...       ...    ...   ...         ...   \n",
       "702477  chrY     CDS  57190204  57190375      +     1  uc004fxm.2   \n",
       "702478  chrY     CDS  57190969  57191085      +     0  uc004fxm.2   \n",
       "702479  chrY     CDS  57191798  57191999      +     0  uc004fxm.2   \n",
       "702480  chrY     CDS  57192603  57192708      +     2  uc004fxm.2   \n",
       "702481  chrY     CDS  57196336  57196369      +     1  uc004fxm.2   \n",
       "\n",
       "       transcript_id exon_number        exon_id  \\\n",
       "0         uc001aal.1           1   uc001aal.1.1   \n",
       "1         uc057auo.1           1   uc057auo.1.1   \n",
       "2         uc057auo.1           2   uc057auo.1.2   \n",
       "3         uc057auo.1           3   uc057auo.1.3   \n",
       "4         uc057aup.1           1   uc057aup.1.1   \n",
       "...              ...         ...            ...   \n",
       "702477    uc004fxm.2           5   uc004fxm.2.5   \n",
       "702478    uc004fxm.2           6   uc004fxm.2.6   \n",
       "702479    uc004fxm.2           7   uc004fxm.2.7   \n",
       "702480    uc004fxm.2           8   uc004fxm.2.8   \n",
       "702481    uc004fxm.2           9   uc004fxm.2.9   \n",
       "\n",
       "                                                nmd_score alignID.1  \n",
       "0        -0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0....  uc001aal  \n",
       "1        0.23,0.23,0.23,0.24,0.24,0.24,0.24,0.24,0.24,...  uc057auo  \n",
       "2        0.31,0.31,0.31,0.31,0.32,0.32,0.32,0.32,0.32,...  uc057auo  \n",
       "3        0.17,0.17,0.17,0.17,0.18,0.17,0.17,0.18,0.18,...  uc057auo  \n",
       "4        0.1,0.1,0.1,0.07,0.08,0.08,0.08,0.08,0.09,0.1...  uc057aup  \n",
       "...                                                   ...       ...  \n",
       "702477   0.58,0.58,0.58,0.61,0.61,0.62,0.61,0.61,0.61,...  uc004fxm  \n",
       "702478   0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,...  uc004fxm  \n",
       "702479   0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.57,...  uc004fxm  \n",
       "702480   0.6,0.65,0.66,0.64,0.59,0.56,0.56,0.54,0.54,0...  uc004fxm  \n",
       "702481   0.05,0.06,0.06,0.06,0.06,0.06,0.06,0.06,0.06,...  uc004fxm  \n",
       "\n",
       "[702482 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Import the table with NMD scores\n",
    "\n",
    "nmdA_df = pd.read_table(\"/workspace/users/msanchezg/notebooks/protein_degradation/NMDetective/nmd.scores.A.tsv.gz\", sep='\\t')\n",
    "nmdA_df['alignID.1'] = nmdA_df['transcript_id'].str.split('.',expand=True)[0]\n",
    "nmdA_df['alignID.1'] = nmdA_df['alignID.1'].str.split(' ',expand=True)[1]\n",
    "nmdA_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript_ens</th>\n",
       "      <th>alignID.1</th>\n",
       "      <th>exonCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328</td>\n",
       "      <td>uc286dmu</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305</td>\n",
       "      <td>uc286dmv</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>uc286dmw</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000619216</td>\n",
       "      <td>uc031tla</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000473358</td>\n",
       "      <td>uc057aty</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247536</th>\n",
       "      <td>ENST00000615362</td>\n",
       "      <td>uc286tlf</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247537</th>\n",
       "      <td>ENST00000617983</td>\n",
       "      <td>uc062frj</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247538</th>\n",
       "      <td>ENST00000613204</td>\n",
       "      <td>uc032quk</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247539</th>\n",
       "      <td>ENST00000615165</td>\n",
       "      <td>uc032qun</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247540</th>\n",
       "      <td>ENST00000621424</td>\n",
       "      <td>uc032qum</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>247380 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Transcript_ens alignID.1  exonCount\n",
       "0       ENST00000456328  uc286dmu          3\n",
       "1       ENST00000450305  uc286dmv          6\n",
       "2       ENST00000488147  uc286dmw         11\n",
       "3       ENST00000619216  uc031tla          1\n",
       "4       ENST00000473358  uc057aty          3\n",
       "...                 ...       ...        ...\n",
       "247536  ENST00000615362  uc286tlf          2\n",
       "247537  ENST00000617983  uc062frj          1\n",
       "247538  ENST00000613204  uc032quk          5\n",
       "247539  ENST00000615165  uc032qun         14\n",
       "247540  ENST00000621424  uc032qum         15\n",
       "\n",
       "[247380 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Import table with UCSC / ENS annotation (UCSC table browser)\n",
    "\n",
    "annot_df = pd.read_table(\"/workspace/users/msanchezg/notebooks/protein_degradation/NMDetective/annotation.ucsc.ens.tsv.gz\", sep=\"\\t\")\n",
    "annot_df['alignID.1'] = annot_df['alignID'].str.split('.',expand=True)[0]\n",
    "annot_df['Transcript_ens'] = annot_df['#name'].str.split('.',expand=True)[0]\n",
    "annot_df = annot_df[['Transcript_ens','alignID.1','exonCount']]\n",
    "annot_df.drop_duplicates(inplace=True, keep='first')\n",
    "annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exonCount</th>\n",
       "      <th>Transcript_ens</th>\n",
       "      <th>nmd_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69091</td>\n",
       "      <td>70005</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>-0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450743</td>\n",
       "      <td>451678</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENST00000426406</td>\n",
       "      <td>0.12,0.12,0.12,0.1,0.11,0.11,0.1,0.11,0.12,0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENST00000332831</td>\n",
       "      <td>0.12,0.12,0.12,0.1,0.11,0.11,0.1,0.11,0.12,0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>925942</td>\n",
       "      <td>926013</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENST00000420190</td>\n",
       "      <td>0.3,0.3,0.3,0.3,0.3,0.3,0.3,0.3,0.3,0.29,0.29...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>chr1</td>\n",
       "      <td>930155</td>\n",
       "      <td>930336</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENST00000420190</td>\n",
       "      <td>0.33,0.33,0.33,0.33,0.33,0.33,0.33,0.33,0.34,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702477</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57190204</td>\n",
       "      <td>57190375</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENST00000369423</td>\n",
       "      <td>0.58,0.58,0.58,0.61,0.61,0.62,0.61,0.61,0.61,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702478</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57190969</td>\n",
       "      <td>57191085</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENST00000369423</td>\n",
       "      <td>0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702479</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57191798</td>\n",
       "      <td>57191999</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENST00000369423</td>\n",
       "      <td>0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.57,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702480</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57192603</td>\n",
       "      <td>57192708</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENST00000369423</td>\n",
       "      <td>0.6,0.65,0.66,0.64,0.59,0.56,0.56,0.54,0.54,0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702481</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57196336</td>\n",
       "      <td>57196369</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9.0</td>\n",
       "      <td>ENST00000369423</td>\n",
       "      <td>0.05,0.06,0.06,0.06,0.06,0.06,0.06,0.06,0.06,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>685197 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM     Start       End Strand Frame exon_number  exonCount  \\\n",
       "0       chr1     69091     70005      +     0           1        3.0   \n",
       "40      chr1    450743    451678      -     0           1        1.0   \n",
       "41      chr1    685719    686654      -     0           1        1.0   \n",
       "42      chr1    925942    926013      +     0           2        7.0   \n",
       "43      chr1    930155    930336      +     0           3        7.0   \n",
       "...      ...       ...       ...    ...   ...         ...        ...   \n",
       "702477  chrY  57190204  57190375      +     1           5        9.0   \n",
       "702478  chrY  57190969  57191085      +     0           6        9.0   \n",
       "702479  chrY  57191798  57191999      +     0           7        9.0   \n",
       "702480  chrY  57192603  57192708      +     2           8        9.0   \n",
       "702481  chrY  57196336  57196369      +     1           9        9.0   \n",
       "\n",
       "         Transcript_ens                                          nmd_score  \n",
       "0       ENST00000641515   -0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0.09,-0....  \n",
       "40      ENST00000426406   0.12,0.12,0.12,0.1,0.11,0.11,0.1,0.11,0.12,0....  \n",
       "41      ENST00000332831   0.12,0.12,0.12,0.1,0.11,0.11,0.1,0.11,0.12,0....  \n",
       "42      ENST00000420190   0.3,0.3,0.3,0.3,0.3,0.3,0.3,0.3,0.3,0.29,0.29...  \n",
       "43      ENST00000420190   0.33,0.33,0.33,0.33,0.33,0.33,0.33,0.33,0.34,...  \n",
       "...                 ...                                                ...  \n",
       "702477  ENST00000369423   0.58,0.58,0.58,0.61,0.61,0.62,0.61,0.61,0.61,...  \n",
       "702478  ENST00000369423   0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,0.61,...  \n",
       "702479  ENST00000369423   0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.58,0.57,...  \n",
       "702480  ENST00000369423   0.6,0.65,0.66,0.64,0.59,0.56,0.56,0.54,0.54,0...  \n",
       "702481  ENST00000369423   0.05,0.06,0.06,0.06,0.06,0.06,0.06,0.06,0.06,...  \n",
       "\n",
       "[685197 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Merge tables\n",
    "\n",
    "nmdA_annot_df = pd.merge(nmdA_df,annot_df, how='left')\n",
    "nmdA_annot_df = nmdA_annot_df[['CHROM','Start','End','Strand','Frame','exon_number','exonCount','Transcript_ens','nmd_score']][~nmdA_annot_df['Transcript_ens'].isnull()]\n",
    "nmdA_annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save table for posterior analysis with pyhton/qmap script\n",
    "\n",
    "nmdA_annot_df.to_csv(r'nmdA.annot.enst.tsv', header=True, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcript ID of genes with ns and fs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c89914b06df14cf9a438b71ba5c0210e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Load CPTAC data', max=7.0, style=ProgressStyle(descriptio…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "#Load CPTAC dataset\n",
    "\n",
    "CPTAC = ['BRCA', 'CCRCC', 'COAD', 'GBM', 'LUAD', 'OV', 'UCEC']\n",
    "cptac_df = pd.DataFrame(columns=['gene', 'cpct_aliquot', 'protein_expression', 'aliquot', 'sample',\n",
    "       'Type', 'median', 'stdev', 'norm_protein_expression', 'log2fpkm',\n",
    "       'log10fpkm', '#Uploaded_variation', 'Location', 'Allele', 'Feature',\n",
    "       'protein_mutation', 'Protein_position', 'Amino_acids', 'Consequence',\n",
    "       'Phenotype', 'Ubiquitinases_Mutated', 'Altered_E3_Ligases',\n",
    "       'Raw_Residual', 'Stability_Change', 'ABS_Stability_Change'])\n",
    "for cancer in tqdm(CPTAC, total=len(CPTAC),desc='Load CPTAC data'):\n",
    "    df = pd.read_csv(\"/workspace/projects/cptac_analysis/data/\"+cancer+\"/dataset_irls.gz\", sep='\\t')\n",
    "    Dataset = [cancer] * len(df)\n",
    "    df['Dataset'] = Dataset\n",
    "    frames = [cptac_df, df]\n",
    "    cptac_df = pd.concat(frames)\n",
    "cptac_df.drop_duplicates(keep='first',inplace=True) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load CCLE dataset\n",
    "\n",
    "ccle_df = pd.read_csv(\"/workspace/projects/cptac_analysis/data/CCLE/dataset_irls.gz\", sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_ns_fs_cptac_df = cptac_df[['Feature']][(cptac_df['Phenotype']=='stop_gained')|(cptac_df['Phenotype']=='frameshift_variant')]\n",
    "genes_ns_fs_ccle_df = ccle_df[['Feature']][(ccle_df['Phenotype']=='stop_gained')|(ccle_df['Phenotype']=='frameshift_variant')]\n",
    "\n",
    "genes_ns_fs_df = pd.concat([genes_ns_fs_cptac_df, genes_ns_fs_ccle_df], ignore_index=True)\n",
    "genes_ns_fs_df.drop_duplicates(inplace=True,keep='first',ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000645829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000645237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000376142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000616317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000642319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8391</th>\n",
       "      <td>ENST00000376670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8392</th>\n",
       "      <td>ENST00000345034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8393</th>\n",
       "      <td>ENST00000273371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8394</th>\n",
       "      <td>ENST00000378076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8395</th>\n",
       "      <td>ENST00000222598</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8396 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Feature\n",
       "0     ENST00000645829\n",
       "1     ENST00000645237\n",
       "2     ENST00000376142\n",
       "3     ENST00000616317\n",
       "4     ENST00000642319\n",
       "...               ...\n",
       "8391  ENST00000376670\n",
       "8392  ENST00000345034\n",
       "8393  ENST00000273371\n",
       "8394  ENST00000378076\n",
       "8395  ENST00000222598\n",
       "\n",
       "[8396 rows x 1 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ns_fs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_ns_fs_df.to_csv(r'genes_ns_fs.tsv', header=True, index=None, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
